{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calephelis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Coperation with Jean-Michel, virus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### find virus related sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "import pandas as pd\n",
    "import os\n",
    "import glob\n",
    "\n",
    "folderBioHPC = '/work/archive/biophysics/Nick_lab/s185491/Celastrina/fq'\n",
    "folder_fq = '/archive/butterfly/ready_fastq/'\n",
    "\n",
    "file_summary = '/home/xcao/w/2018Calephelis/summary/20181217Calephelis.xlsx'\n",
    "df_summary = pd.read_excel(file_summary,sheet_name='Sheet1')\n",
    "sample_prefix = [e[1:] for e in df_summary['ID']]\n",
    "print('unique sample prefix', len(set(sample_prefix)))\n",
    "\n",
    "files_fq = glob.glob(folder_fq+'*')\n",
    "files_use = [e for e in files_fq if os.path.basename(e).split('_')[0] in sample_prefix]\n",
    "prefix_fq = [os.path.basename(e).split('_')[0] for e in files_use]\n",
    "\n",
    "sample_prefix =set(prefix_fq)\n",
    "\n",
    "txt = '''cd {workingfolder} && mkdir {name} && cd {name} && /home/xcao/p/mira/mira_4.0.2_linux-gnu_x86_64_static/bin/mirabait -k 21 {baitTarget} /archive/butterfly/ready_fastq/{name}_R1.fastq {name}_R1.fastq && /home/xcao/p/mira/mira_4.0.2_linux-gnu_x86_64_static/bin/mirabait -k 21 {baitTarget} /archive/butterfly/ready_fastq/{name}_R2.fastq {name}_R2.fastq && /home/xcao/p/SPAdes/SPAdes-3.13.0-Linux/bin/spades.py -t 4 --s1 {name}_R1.fastq --s2 {name}_R2.fastq -o ./spades/ && mv ./spades/contigs.fasta ../{name}_contig.fa && rm -rf {workingfolder}{name}/spades && rm {workingfolder}{name}/hashstat.bin\n",
    "'''\n",
    "\n",
    "workingfolder = '/home/xcao/w/2018Calephelis/20181217virus/scaffold1380/'\n",
    "baitTarget = '/home/xcao/w/2018Calephelis/20181217virus/scaffold1380/scaffold1308.fa'\n",
    "open(workingfolder+'jobs.cmds','w').write(''.join(txt.format(workingfolder=workingfolder, baitTarget=baitTarget, name = name) for name in sample_prefix))\n",
    "\n",
    "baitTarget='/home/xcao/w/2018Calephelis/20181217virus/scaffold6262/scaffold6262.fa'\n",
    "workingfolder='/home/xcao/w/2018Calephelis/20181217virus/scaffold6262/'\n",
    "open(workingfolder+'jobs.cmds','w').write(''.join(txt.format(workingfolder=workingfolder, baitTarget=baitTarget, name = name) for name in sample_prefix))\n",
    "\n",
    "#nohup python /home/xcao/p/xiaolongTools/multiThread.py 48 /home/xcao/w/2018Calephelis/20181217virus/scaffold1380/jobs.cmds &\n",
    "#nohup python /home/xcao/p/xiaolongTools/multiThread.py 48 /home/xcao/w/2018Calephelis/20181217virus/scaffold6262/jobs.cmds &\n",
    "\n",
    "folder = '/home/xcao/w/2018Calephelis/20181217virus/scaffold6262/'\n",
    "contigs_prefix = glob.glob(folder+'*_contig.fa')\n",
    "contigs_prefix = [os.path.basename(e).split('_')[0] for e in contigs_prefix]\n",
    "print([e for e in sample_prefix if e not in contigs_prefix])\n",
    "#the missing ones are ['3505', '3540', '3585', '3847']\n",
    "missing_prefix = [e for e in sample_prefix if e not in contigs_prefix]\n",
    "baitTarget='/home/xcao/w/2018Calephelis/20181217virus/scaffold6262/scaffold6262.fa'\n",
    "workingfolder='/home/xcao/w/2018Calephelis/20181217virus/scaffold6262/'\n",
    "txt = '''cd {workingfolder} && rm -rf {workingfolder}{name} && mkdir {name} && cd {name} && /home/xcao/p/mira/mira_4.0.2_linux-gnu_x86_64_static/bin/mirabait -k 21 {baitTarget} /archive/butterfly/ready_fastq/{name}_400_R1.fastq {name}_R1.fastq && /home/xcao/p/mira/mira_4.0.2_linux-gnu_x86_64_static/bin/mirabait -k 21 {baitTarget} /archive/butterfly/ready_fastq/{name}_400_R2.fastq {name}_R2.fastq && /home/xcao/p/SPAdes/SPAdes-3.13.0-Linux/bin/spades.py -t 4 --s1 {name}_R1.fastq --s2 {name}_R2.fastq -o ./spades/ && mv ./spades/contigs.fasta ../{name}_contig.fa && rm -rf {workingfolder}{name}/spades && rm {workingfolder}{name}/hashstat.bin\n",
    "'''\n",
    "for name in missing_prefix:\n",
    "    print(txt.format(workingfolder=workingfolder, baitTarget=baitTarget, name = name))\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## summary"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### check existence of files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### fastq\n",
    "```\n",
    "import pandas as pd\n",
    "import os\n",
    "import glob\n",
    "\n",
    "folderBioHPC = '/work/archive/biophysics/Nick_lab/s185491/Celastrina/fq'\n",
    "folder_fq = '/archive/butterfly/ready_fastq/'\n",
    "folder_summary = '/home/xcao/w/2018Calephelis/summary/'\n",
    "\n",
    "file_summary = os.path.join(folder_summary,'20181217Calephelis.xlsx')\n",
    "df_summary = pd.read_excel(file_summary,sheet_name='Sheet1')\n",
    "sample_prefix = [e[1:] for e in df_summary['ID']]\n",
    "print('unique sample prefix', len(set(sample_prefix)))\n",
    "\n",
    "files_fq = glob.glob(folder_fq+'*')\n",
    "files_use = [e for e in files_fq if os.path.basename(e).split('_')[0] in sample_prefix]\n",
    "\n",
    "files_possible = []\n",
    "for e in files_fq:\n",
    "    if any([os.path.basename(e).startswith(s) for s in sample_prefix]):\n",
    "        if e not in files_use:\n",
    "            files_possible.append(e)\n",
    "print(files_possible,'possible abnormal name, if not sampleID_R1/R2.fastq')\n",
    "\n",
    "dc_fq=dict([e,[]] for e in sample_prefix)\n",
    "for e in files_use:\n",
    "    dc_fq[os.path.basename(e).split('_')[0]].append(e)\n",
    "\n",
    "sample_fq = [e for e,v in dc_fq.items() if len(v)==3]\n",
    "open(os.path.join(folder_summary, 'samples_with_fq.txt'),'w').write('\\n'.join(sample_fq))\n",
    "\n",
    "from collections import Counter\n",
    "print(Counter([len(e) for e in dc_fq.values()]))\n",
    "```\n",
    "\n",
    "284 unique sample names, 266 of them have good fastq files. 18 have no fastq files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### map files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "#check existance of map files\n",
    "files_map = glob.glob('/archive/butterfly/maps/debiased/*.map')\n",
    "files_map = set([os.path.basename(e) for e in files_map])\n",
    "print('number of map files',len(files_map))\n",
    "sample_map = [e for e in files_map if os.path.basename(e).split('_')[0] in sample_prefix]\n",
    "print(len(sample_map),'map files use')\n",
    "\n",
    "dc_map = dict([e,[]] for e in sample_prefix)\n",
    "for e in sample_map:\n",
    "    dc_map[os.path.basename(e).split('_')[0]].append(e)\n",
    "\n",
    "print(Counter([len(e) for e in dc_map.values()]))\n",
    "#get reference genomes\n",
    "ls_reference = []\n",
    "for e in sample_map:\n",
    "    ls_reference.append(os.path.basename(e).split('_',1)[1])\n",
    "print('reference Counter\\n',Counter(ls_reference))\n",
    "\n",
    "#use 3574_assembly_v1_withMito_snp_step2.map as reference\n",
    "ref_genome = '3574_assembly_v1_withMito'\n",
    "sample_map = [e for e in sample_prefix if e+'_'+ref_genome+'_snp_step2.map' in files_map]\n",
    "files_use_map = ['/archive/butterfly/maps/debiased/' + e+'_'+ref_genome+'_snp_step2.map' for e in sample_map]\n",
    "print(len(sample_map),'map files use')\n",
    "open(os.path.join(folder_summary, 'mapfiles.txt'),'w').write('\\n'.join(files_use_map))\n",
    "\n",
    "print(len(set(sample_fq + sample_map)))\n",
    "#all sample_map is in sample_fq\n",
    "sample_todo = [e for e in sample_fq if e not in sample_map]\n",
    "print([e for e in sample_fq if e not in sample_map])\n",
    "\n",
    "#check existance of statistics files\n",
    "files = glob.glob('/archive/butterfly/unbias_pipeline_info/step4_postprocessing/final_stats/*.report')\n",
    "files = set([os.path.basename(e) for e in files])\n",
    "sample_stats = [e for e in sample_prefix if e+'_'+ref_genome+'_stat.report' in files]\n",
    "print(len(sample_stats),'file stats')\n",
    "print('files missing',[e for e in sample_prefix if e not in sample_map])\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "after checking with Qian, 3574 is a better reference genome.\n",
    "20181224 checked. All 266 file exists"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### summarize stat files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "ls_file = ['/archive/butterfly/unbias_pipeline_info/step4_postprocessing/final_stats/' + e+'_'+ref_genome+'_stat.report' for e in sample_prefix if e+'_'+ref_genome+'_stat.report' in files]\n",
    "import pandas as pd\n",
    "df_align_stat = pd.concat([pd.read_csv(e,sep='\\t',dtype=str) for e in ls_file],ignore_index=True)\n",
    "df_align_stat['sample'] =df_align_stat['sample'].apply(lambda x:'S'+x)\n",
    "print(df_align_stat.shape)\n",
    "df_align_stat.head()\n",
    "df_align_stat.to_csv('/home/xcao/w/2018Calephelis/summary/20181225Calephelis_align_stat.csv',sep='\\t')\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### map files to npInt8"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "files_map = '/home/xcao/w/2018Calephelis/summary/mapfiles.txt'\n",
    "outfolder1 = '/home/xcao/w/2018Calephelis/summary/map.npInt8.0'\n",
    "outfolder2 = '/home/xcao/w/2018Calephelis/summary/map.npInt8.2'\n",
    "\n",
    "import sys\n",
    "sys.path.append('/home/xcao/p/xiaolongTools/utils/mapFile')\n",
    "import mapfileIO\n",
    "mapfileIO.read2Int8s(filenames=files_map, strand=0, changeGap=True, threads=32, outfolder=outfolder1)\n",
    "mapfileIO.read2Int8s(filenames=files_map, strand=2, changeGap=True, threads=32, outfolder=outfolder2)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Count gaps in each position"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "import sys\n",
    "sys.path.append('/home/xcao/p/xiaolongTools/utils/mapFile')\n",
    "import mapfileIO\n",
    "import glob\n",
    "import numpy as np\n",
    "import pickle\n",
    "\n",
    "folder = '/home/xcao/w/2018Calephelis/summary/map.npInt8.0/'\n",
    "outfile = '/home/xcao/w/2018Calephelis/summary/map.sitesInfo/gapCount266'\n",
    "outfileHist = outfile + '.hist'\n",
    "files = glob.glob(folder+'*')\n",
    "print('total map files', len(files))\n",
    "\n",
    "#ls_mapInt8 = mapfileIO.loadMapBinaries(files, threads=8)#multithreading not working. use the slow method\n",
    "ls_mapInt8 = [mapfileIO.loadMapBinary(f) for f in files]\n",
    "temp = ls_mapInt8[0]\n",
    "maplen = len(temp)\n",
    "gapCount = np.zeros(maplen,dtype=np.int16)\n",
    "for e in ls_mapInt8:\n",
    "    gapCount += (e==45)\n",
    "\n",
    "with open(outfile,'wb') as f:\n",
    "    pickle.dump(gapCount,f)\n",
    "\n",
    "gapHist = np.unique(gapCount,return_counts=True)\n",
    "with open(outfileHist,'w') as f:\n",
    "    for c, p in zip(*gapHist):\n",
    "        f.write('{}\\t{}\\n'.format(c,p))\n",
    "\n",
    "#average gap per site\n",
    "gapAvg = gapCount.sum() / maplen\n",
    "print('average gap per site', gapAvg)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### get some information about the genome"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### genome header, scaffold length"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "filename = '/home/xcao/w/genomes/Calephelis/3574_assembly_v1_withMito.fa'\n",
    "from Bio import SeqIO\n",
    "fout = open(filename + '.head','w')\n",
    "for _e in SeqIO.parse(filename,'fasta'):\n",
    "    fout.write(_e.description+'\\n')\n",
    "fout.close()\n",
    "fout = open(filename +'.len','w')\n",
    "for _e in SeqIO.parse(filename,'fasta'):\n",
    "    fout.write(_e.id+'\\t'+str(len(_e.seq))+'\\n')\n",
    "fout.close()\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### positions of mitochondria, mRNA, CDS\n",
    "for the gff file by Qian, there are only mRNA, CDS. So we can group the regions in the genome to 4 groups:  \n",
    "mitochondria: 1  #Note: 20181226: just found that though the reference contains mitochondrial sequences, the map file does not contain.\n",
    "intron : 2  \n",
    "CDS/exon: 6  \n",
    "inter-gene: 0  \n",
    "intron+CDS forms gene_region: 2 or 6\n",
    "\n",
    "```\n",
    "file_gff = '/home/xcao/w/genomes/Calephelis/3574_assembly_v1.gff'\n",
    "file_genome_len = '/home/xcao/w/genomes/Calephelis/3574_assembly_v1_withMito.fa.len'\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pickle\n",
    "\n",
    "df_scf = pd.read_csv(file_genome_len,sep='\\t', header=None)\n",
    "df_scf.columns = ['scf','scf_len']\n",
    "df_scf = df_scf.iloc[:-1] #remove the mitochondrial row\n",
    "genomelen = df_scf['scf_len'].sum()\n",
    "\n",
    "df_scf['start'] = 0\n",
    "df_scf['end'] = df_scf['scf_len']\n",
    "for i in df_scf.index[1:]:\n",
    "    df_scf.loc[i,'start'] = df_scf.loc[i-1,'start'] + df_scf.loc[i-1, 'scf_len']\n",
    "    df_scf.loc[i,'end'] = df_scf.loc[i,'start'] + df_scf.loc[i,'scf_len']\n",
    "\n",
    "dc_scf2start = dict(zip(df_scf['scf'], df_scf['start']))\n",
    "dc_scf2end = dict(zip(df_scf['scf'], df_scf['end']))\n",
    "\n",
    "df_gff = pd.read_csv(file_gff,sep='\\t', header=None)\n",
    "df_gff.columns = ['scf', 'program', 'type', 'start', 'end', 'special', 'strand','exonN','transcript_id']\n",
    "df_gff = df_gff.set_index('scf')\n",
    "\n",
    "gffnp = np.zeros(genomelen,dtype=np.uint8)\n",
    "##set mitochondrial sites to 1\n",
    "#start = dc_scf2start['3574_mito']\n",
    "#end = dc_scf2end['3574_mito']\n",
    "#gffnp[range(start,end)] = 1\n",
    "#set mRNA region to 2\n",
    "for scf, scf_gff in df_gff.iterrows():\n",
    "    start = scf_gff['start']-1\n",
    "    end = scf_gff['end']\n",
    "    start += dc_scf2start[scf]\n",
    "    end += dc_scf2start[scf]\n",
    "    if scf_gff['type'] == 'mRNA':\n",
    "        gffnp[range(start,end)] = 2\n",
    "#set CDS region to 6\n",
    "for scf, scf_gff in df_gff.iterrows():\n",
    "    start = scf_gff['start']-1\n",
    "    end = scf_gff['end']\n",
    "    start += dc_scf2start[scf]\n",
    "    end += dc_scf2start[scf]\n",
    "    if scf_gff['type'] == 'CDS':\n",
    "        gffnp[range(start,end)] = 6\n",
    "\n",
    "with open('/home/xcao/w/2018Calephelis/summary/map.sitesInfo/gff.0inter.2intron.6CDS','wb') as f:\n",
    "    pickle.dump(gffnp,f)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### get location of exons in CDS alignment\n",
    "```\n",
    "file_gff = '/home/xcao/w/genomes/Calephelis/3574_assembly_v1.gff'\n",
    "file_genome_len = '/home/xcao/w/genomes/Calephelis/3574_assembly_v1_withMito.fa.len'\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pickle\n",
    "\n",
    "df_scf = pd.read_csv(file_genome_len,sep='\\t', header=None)\n",
    "df_scf.columns = ['scf','scf_len']\n",
    "df_scf = df_scf.iloc[:-1] #remove the mitochondrial row\n",
    "genomelen = df_scf['scf_len'].sum()\n",
    "\n",
    "df_scf['start'] = 0\n",
    "df_scf['end'] = df_scf['scf_len']\n",
    "for i in df_scf.index[1:]:\n",
    "    df_scf.loc[i,'start'] = df_scf.loc[i-1,'start'] + df_scf.loc[i-1, 'scf_len']\n",
    "    df_scf.loc[i,'end'] = df_scf.loc[i,'start'] + df_scf.loc[i,'scf_len']\n",
    "\n",
    "dc_scf2start = dict(zip(df_scf['scf'], df_scf['start']))\n",
    "dc_scf2end = dict(zip(df_scf['scf'], df_scf['end']))\n",
    "\n",
    "df_gff = pd.read_csv(file_gff,sep='\\t', header=None)\n",
    "df_gff.columns = ['scf', 'program', 'type', 'start', 'end', 'special', 'strand','phase','transcript_id']\n",
    "df_gff['map_start'] = df_gff.apply(lambda x:dc_scf2start[x['scf']] + x['start']-1, axis = 'columns')\n",
    "df_gff['map_end'] = df_gff.apply(lambda x:dc_scf2start[x['scf']] + x['end'], axis = 'columns')\n",
    "\n",
    "with open('/home/xcao/w/2018Calephelis/summary/map.sitesInfo/gff.0inter.2intron.6CDS','rb') as f:\n",
    "    gffnp = pickle.load(f)\n",
    "\n",
    "df_exon = df_gff[df_gff['type'] == 'CDS'].copy()\n",
    "\n",
    "\n",
    "npExonDC = -np.ones(genomelen, dtype=np.int64)#the same length as the genome.\n",
    "start = 0\n",
    "for n,v in enumerate(gffnp):\n",
    "    if v == 6:\n",
    "        npExonDC[n] = start\n",
    "        start += 1\n",
    "\n",
    "df_exon['cds_start'] = df_exon['map_start'].apply(lambda x: npExonDC[x])\n",
    "df_exon['cds_end'] = df_exon['map_end'].apply(lambda x: npExonDC[x-1]+1)\n",
    "df_exon['exon_N'] = df_exon.groupby(['transcript_id']).cumcount()+1\n",
    "\n",
    "fout = open('/home/xcao/w/genomes/Calephelis/3574_assembly_v1_withMito.alignment.exon.loc','w')\n",
    "for _n, row in df_exon.iterrows():\n",
    "    fout.write('{transcript_id}.e{exon_n}\\t{cds_start}\\t{cds_end}\\n'.format(transcript_id = row['transcript_id'], exon_n = row['exon_N'], cds_start = row['cds_start'], cds_end = row['cds_end']))\n",
    "fout.close()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### split the genome to fragments\n",
    "\n",
    "split the genome according to different scan window. get a dictionary with name as key and np array of position in genome as value\n",
    "\n",
    "\n",
    "12140 fragments for window_size 50k  \n",
    "4893 fragments for window_size 100k  \n",
    "\n",
    "```\n",
    "file_genome_info = '/home/xcao/w/genomes/Calephelis/3574_assembly_v1_withMito.fa.len'\n",
    "window_size = 50000 #split the genome to parts >= window_size but < 2*window_size. \n",
    "fout = '/home/xcao/w/2018Calephelis/summary/map.sitesInfo/dc_sites_window50000'\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pickle\n",
    "\n",
    "df_genome = pd.read_csv(file_genome_info,sep='\\t',header=None)\n",
    "df_genome.columns = ['scf','scf_len']\n",
    "df_genome = df_genome[:-1] #drop the row with mitochondria\n",
    "print('genome size is', df_genome['scf_len'].sum()) #genome size is 808598630\n",
    "\n",
    "#get start and end of each scaffold in map file\n",
    "starts = [0]\n",
    "for scf_len in df_genome['scf_len']:\n",
    "    starts.append(starts[-1] + scf_len)\n",
    "df_genome['start'] = starts[:-1]\n",
    "df_genome['end'] = starts[1:]\n",
    "\n",
    "#filter df_genome, keep those only larger than window_size\n",
    "df_filter = df_genome[df_genome['scf_len']>=window_size] #3733 outof 40395 scaffolds left\n",
    "\n",
    "#determine th splittig parts for each scaffold\n",
    "df_filter['parts'] = df_filter['scf_len'] // window_size\n",
    "\n",
    "#generate a dictionary with sequence name and locations in map file\n",
    "dc_frags = {}\n",
    "for n, row in df_filter.iterrows():\n",
    "    scf_positions = np.array(range(row['start'], row['end']), dtype=np.int32)\n",
    "    ls_scf_positions = np.array_split(scf_positions,row['parts'])\n",
    "    for frag_n, frag_position in enumerate(ls_scf_positions):\n",
    "        frag_id = row['scf'] + '_' +str(frag_n)\n",
    "        dc_frags[frag_id] = frag_position\n",
    "\n",
    "with open(fout,'wb') as f:\n",
    "    pickle.dump(dc_frags, f)\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Phylogenetic Trees"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### identify z-chromsomes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### run promer3\n",
    "```\n",
    "workfolder = '/home/xcao/w/2018Calephelis/20181225zchromosome/'\n",
    "target = 'Calephelis3574'\n",
    "target_genome = '/home/xcao/w/genomes/Calephelis/3574_assembly_v1_withMito.fa'\n",
    "ls_reference = ['Spodopetera', 'Heliconius','Bombyx']\n",
    "ls_reference_genome=['/home/xcao/w/genomes/Spodoptera_litura/GCF_002706865.1_ASM270686v1_genomic.fna', '/home/xcao/w/genomes/Heliconius_melpomene/Heliconius_melpomene_melpomene_Hmel2.5.scaffolds.fa', '/home/xcao/w/genomes/Bombyx_mori/Bomo_genome_assembly.fa']\n",
    "\n",
    "commands = '''\n",
    "cd {workfolder} &&\\\n",
    "/home/xcao/p/mummer/MUMmer3.23/promer  -p {target}_2_{reference}.promer3 {reference_genome} {target_genome} &&\\\n",
    "/home/xcao/p/mummer/MUMmer3.23/delta-filter -q -r {target}_2_{reference}.promer3.delta >{target}_2_{reference}.promer3.delta.qr &&\\\n",
    "/home/xcao/p/mummer/MUMmer3.23/mummerplot {target}_2_{reference}.promer3.delta.qr -t postscript -p {target}_2_{reference}.promer3.delta.qr &&\\\n",
    "/home/xcao/p/mummer/MUMmer3.23/show-coords -rcl {target}_2_{reference}.promer3.delta.qr >{target}_2_{reference}.promer3.coords.qr &\n",
    "'''\n",
    "\n",
    "for reference, reference_genome in zip(ls_reference, ls_reference_genome):\n",
    "    print(commands.format(reference=reference,reference_genome=reference_genome,target=target,target_genome=target_genome,workfolder=workfolder))\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####  assign scaffold of Celastrina according to 'Spodopetera', 'Heliconius','Bombyx'\n",
    "\n",
    "```\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "files = '''/home/xcao/w/2018Calephelis/20181225zchromosome/Calephelis3574_2_Bombyx.promer3.coords.qr\n",
    "/home/xcao/w/2018Calephelis/20181225zchromosome/Calephelis3574_2_Heliconius.promer3.coords.qr\n",
    "/home/xcao/w/2018Calephelis/20181225zchromosome/Calephelis3574_2_Spodopetera.promer3.coords.qr\n",
    "'''.split()\n",
    "filename_coord = files[2]\n",
    "\n",
    "filename = filename_coord\n",
    "l = open(filename).readlines()\n",
    "ls =[e.split() for e in l[5:]]\n",
    "\n",
    "df_coord = pd.DataFrame(ls)\n",
    "df_coord = df_coord[[0,1,3,4,6,7,9,10,11,13,14,16,17,19,20,21,22]].copy()\n",
    "df_coord[[0,1,3,4,6,7,9,10,11,13,14,16,17,19,20]] = df_coord[[0,1,3,4,6,7,9,10,11,13,14,16,17,19,20]].apply(pd.to_numeric)\n",
    "df_coord.columns = ['ref_start','ref_end','query_start','query_end','ref_match_len','query_match_len','identity','similarity','percent_stop_codon','ref_len','query_len','ref_coverage','query_coverage','ref_reading_frame','query_reading_frame','ref_id','query_id']\n",
    "df_coord_group_qr = df_coord.groupby(['query_id','ref_id'])\n",
    "\n",
    "df_qr = df_coord_group_qr['ref_match_len'].sum()\n",
    "df_summary = pd.DataFrame([ [e,df_qr[e].idxmax(),df_qr[e].max(),df_qr[e].sum(), df_qr[e].max()/df_qr[e].sum()] for e in df_qr.index.unique(0)])\n",
    "df_summary.columns = ['test_id','ref_id','scaffold_match_len','total_match_len','ratio_specific']\n",
    "\n",
    "\n",
    "for n in range(df_summary.shape[0]):\n",
    "    temp_df = df_coord_group_qr.get_group((df_summary['test_id'][n],df_summary['ref_id'][n]))\n",
    "    match_fragments = temp_df.shape[0]\n",
    "    df_summary.loc[n,'match_fragments'] = match_fragments\n",
    "    positions = []\n",
    "    for m in range(match_fragments):\n",
    "        start = temp_df.iloc[m]['ref_start']\n",
    "        end = temp_df.iloc[m]['ref_end']\n",
    "        if start > end:\n",
    "            start, end = end, start\n",
    "        positions += list(range(start, end))\n",
    "    position_avg = np.median(positions)\n",
    "    df_summary.loc[n,'position_avg'] = position_avg\n",
    "\n",
    "\n",
    "df_summary['match_fragments'] = df_summary['match_fragments'].astype(np.int64)\n",
    "df_summary['position_avg'] = df_summary['position_avg'].astype(np.int64)\n",
    "df_summary = df_summary.sort_values(by=['ref_id','position_avg'])\n",
    "\n",
    "outfile = os.path.join(os.path.dirname(filename_coord), '20181226'+os.path.basename(filename_coord).split('.')[0]+'.csv')\n",
    "df_summary.to_csv(outfile,sep='\\t',index=False)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### get z-chromosome scaffolds according to 'Spodopetera', 'Heliconius','Bombyx'\n",
    "\n",
    "z-chromosome infor based on (scaf num, total length)  \n",
    "Bombyx: 188 36080881  \n",
    "Heliconius: 139 33192451  \n",
    "Spodopetera: 163 34135513  \n",
    "all: 87 29582058  \n",
    "  \n",
    "\n",
    "\n",
    "```\n",
    "import pandas as pd\n",
    "\n",
    "filename = '/home/xcao/w/2018Calephelis/20181225zchromosome/20181226Calephelis3574_2_Bombyx.csv'\n",
    "df_bom = pd.read_csv(filename,sep='\\t')\n",
    "\n",
    "filename = '/home/xcao/w/2018Calephelis/20181225zchromosome/20181226Calephelis3574_2_Heliconius.csv'\n",
    "df_hel = pd.read_csv(filename,sep='\\t')\n",
    "\n",
    "filename = '/home/xcao/w/2018Calephelis/20181225zchromosome/20181226Calephelis3574_2_Spodopetera.csv'\n",
    "df_spo = pd.read_csv(filename,sep='\\t')\n",
    "\n",
    "df_bom_z = df_bom[df_bom['ref_id'] == 'Bomo_Chr1']\n",
    "df_hel_z = df_hel[df_hel['ref_id'] == 'Hmel221001o']\n",
    "df_spo_z = df_spo[df_spo['ref_id'] == 'NC_036217.1']\n",
    "\n",
    "zscf_bom = set(df_bom_z['test_id'])\n",
    "zscf_hel = set(df_hel_z['test_id'])\n",
    "zscf_spo = set(df_spo_z['test_id'])\n",
    "zscf_all = zscf_bom & zscf_hel & zscf_spo\n",
    "\n",
    "file_genome = '/home/xcao/w/genomes/Calephelis/3574_assembly_v1_withMito.fa'\n",
    "from Bio import SeqIO\n",
    "fout = open(file_genome+'.head','w')\n",
    "for _e in SeqIO.parse(file_genome,'fasta'):\n",
    "    fout.write(_e.description+'\\n')\n",
    "fout.close()\n",
    "fout = open(file_genome+'.len','w')\n",
    "for _e in SeqIO.parse(file_genome,'fasta'):\n",
    "    fout.write(_e.id+'\\t'+str(len(_e.seq))+'\\n')\n",
    "fout.close()\n",
    "\n",
    "dc_scaflen = {}\n",
    "filename = file_genome+'.len'\n",
    "for line in open(filename):\n",
    "    scf,l = line.split()\n",
    "    l=int(l)\n",
    "    dc_scaflen[scf] = l\n",
    "\n",
    "print('Bombyx:', len(zscf_bom), sum([dc_scaflen[e] for e in zscf_bom]))\n",
    "print('Heliconius:', len(zscf_hel), sum([dc_scaflen[e] for e in zscf_hel]))\n",
    "print('Spodopetera:', len(zscf_spo), sum([dc_scaflen[e] for e in zscf_spo]))\n",
    "print('all:', len(zscf_all), sum([dc_scaflen[e] for e in zscf_all]))\n",
    "\n",
    "open('/home/xcao/w/2018Calephelis/summary/20181226Calephelis3574_z_scf_basedon3','w').write('\\n'.join(zscf_all))\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### generate a filter np.array file to extract z-chromosome sequences\n",
    "\n",
    "npSites is a numpy array with True and False. If True, means a z-chromosome sites based on the three reference species\n",
    "```\n",
    "file_genomelen = '/home/xcao/w/genomes/Calephelis/3574_assembly_v1_withMito.fa.len'\n",
    "file_targetscf = '/home/xcao/w/2018Calephelis/summary/20181226Calephelis3574_z_scf_basedon3'\n",
    "\n",
    "import numpy as np\n",
    "import pickle\n",
    "import itertools\n",
    "sites = [] #True for keep, False not keep. The same length as the genome\n",
    "target_scfs = open(file_targetscf).read().split()\n",
    "target_scfs = set(target_scfs)\n",
    "ls_genomelen = open(file_genomelen).readlines()[:-1] #remove mitochondrial line\n",
    "for line in ls_genomelen:\n",
    "    scf,scflen = line.split()\n",
    "    scflen = int(scflen)\n",
    "    if scf in target_scfs:\n",
    "        sites.append([True] * scflen)\n",
    "    else:\n",
    "        sites.append([False] * scflen)\n",
    "sites = list(itertools.chain.from_iterable(sites))\n",
    "npSites = np.array(sites)\n",
    "with open('/home/xcao/w/2018Calephelis/summary/map.sitesInfo/zChrFilter','wb') as f:\n",
    "    pickle.dump(npSites,f)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### make z-chromosme phylogenetic tree"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####  extract sequence\n",
    "\n",
    "```\n",
    "import sys\n",
    "sys.path.append('/home/xcao/p/xiaolongTools/utils/mapFile')\n",
    "import mapfileIO\n",
    "import glob\n",
    "\n",
    "folder = '/home/xcao/w/2018Calephelis/summary/map.npInt8.0/'\n",
    "outfile = '/home/xcao/w/2018Calephelis/20181226Phylo/zChrBased3Ref/Calephelis266.zChr'\n",
    "file_filter = '/home/xcao/w/2018Calephelis/summary/map.sitesInfo/zChrFilter'\n",
    "files = glob.glob(folder+'*')\n",
    "print('total map npInt8 files', len(files))\n",
    "\n",
    "mapfileIO.npInt8stoFastas(files, threads=32, output=outfile, headerFun=lambda x:x.split('_')[0], npFilter=file_filter, npFilterFun=None, merged=True)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### filter sequence\n",
    "\n",
    "```\n",
    "#nonGap ratio 0.8\n",
    "python /home/xcao/p/xiaolongTools/utils/fastaAlignmentFilter.py -i /home/xcao/w/2018Calephelis/20181226Phylo/zChrBased3Ref/Calephelis266.zChr -o /home/xcao/w/2018Calephelis/20181226Phylo/zChrBased3Ref/Calephelis266.zChr.filter0.8 -t 32 -b 0 -g 0.8 -B 0\n",
    "#nonGap ratio 0.7\n",
    "python /home/xcao/p/xiaolongTools/utils/fastaAlignmentFilter.py -i /home/xcao/w/2018Calephelis/20181226Phylo/zChrBased3Ref/Calephelis266.zChr -o /home/xcao/w/2018Calephelis/20181226Phylo/zChrBased3Ref/Calephelis266.zChr.filter0.7 -t 32 -b 0 -g 0.7 -B 0\n",
    "#nonGap ratio 0.6\n",
    "python /home/xcao/p/xiaolongTools/utils/fastaAlignmentFilter.py -i /home/xcao/w/2018Calephelis/20181226Phylo/zChrBased3Ref/Calephelis266.zChr -o /home/xcao/w/2018Calephelis/20181226Phylo/zChrBased3Ref/Calephelis266.zChr.filter0.6 -t 32 -b 0 -g 0.6 -B 0\n",
    "#nonGap ratio 0.5\n",
    "python /home/xcao/p/xiaolongTools/utils/fastaAlignmentFilter.py -i /home/xcao/w/2018Calephelis/20181226Phylo/zChrBased3Ref/Calephelis266.zChr -o /home/xcao/w/2018Calephelis/20181226Phylo/zChrBased3Ref/Calephelis266.zChr.filter0.5 -t 32 -b 0 -g 0.5 -B 0\n",
    "#nonGap ratio 0.4\n",
    "python /home/xcao/p/xiaolongTools/utils/fastaAlignmentFilter.py -i /home/xcao/w/2018Calephelis/20181226Phylo/zChrBased3Ref/Calephelis266.zChr -o /home/xcao/w/2018Calephelis/20181226Phylo/zChrBased3Ref/Calephelis266.zChr.filter0.4 -t 32 -b 0 -g 0.4 -B 0\n",
    "#nonGap ratio 0.3\n",
    "python /home/xcao/p/xiaolongTools/utils/fastaAlignmentFilter.py -i /home/xcao/w/2018Calephelis/20181226Phylo/zChrBased3Ref/Calephelis266.zChr -o /home/xcao/w/2018Calephelis/20181226Phylo/zChrBased3Ref/Calephelis266.zChr.filter0.3 -t 32 -b 0 -g 0.3 -B 0\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### run a quick one with filter 0.8 to get the guider tree for examl\n",
    "```\n",
    "txt_qsub = '''#!/bin/bash\n",
    "#SBATCH --job-name=zChr\n",
    "#SBATCH --partition=32GB\n",
    "#SBATCH --nodes=32\n",
    "#SBATCH --ntasks=1024\n",
    "#SBATCH --time=5-00:00:00\n",
    "#SBATCH --output=0.%j.out\n",
    "#SBATCH --mail-user=xiaolong.cao@utsouthwestern.edu\n",
    "#SBATCH --mail-type=ALL\n",
    "\n",
    "module load openmpi/intel/3.1.1\n",
    "\n",
    "/home2/s185491/p/anaconda3/anaconda520/bin/python3 /home2/s185491/p/xiaolongTools/utils/runExaML_withFastaForLargeFiles.py -i {filename} -T 1024\n",
    "'''\n",
    "\n",
    "files = ['/work/biophysics/s185491/20181226Calephelis/Phylo/zChr/Calephelis266.zChr.filter0.8']\n",
    "\n",
    "import os\n",
    "for filename in files:\n",
    "    fout = open('/work/biophysics/s185491/20181226Calephelis/Phylo/zChr/cmds/qsub.Calephelis266.zChr0.8','w')\n",
    "    fout.write(txt_qsub.format(filename=filename,n=os.path.basename(filename)))\n",
    "    fout.close()\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### make mitochondrial tree"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### extract sequences\n",
    "\n",
    "```\n",
    "import sys\n",
    "sys.path.append('/home/xcao/p/xiaolongTools/utils/mapFile')\n",
    "import mapfileIO\n",
    "filename = '/home/xcao/w/2018Calephelis/summary/mapfiles.txt'\n",
    "ls_files = open(filename).read().split()\n",
    "files = [f.replace('3574_assembly_v1_withMito_snp_step2.map', '3574_assembly_v1_withMito_snp_step2_MITO.map') for f in ls_files]\n",
    "outfile = '/home/xcao/w/2018Calephelis/20181226Phylo/mitochondria/Calephelis266.mito'\n",
    "mapfileIO.mapFilestoFastas(files, threads=32, output=outfile, headerFun=lambda x:x.split('/')[-1].split('_')[0], npFilter=None, npFilterFun=None, merged=True)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### run MrBayes\n",
    "\n",
    "cp sequence to bioHPC. For runMrBayes_largeInput, change chain to 2, and number of generation to 2 million\n",
    "```\n",
    "#!/bin/bash\n",
    "#SBATCH --job-name=mito\n",
    "#SBATCH --partition=256GB\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=48\n",
    "#SBATCH --time=5-00:00:00\n",
    "#SBATCH --output=0.%j.out\n",
    "#SBATCH --mail-user=xiaolong.cao@utsouthwestern.edu\n",
    "#SBATCH --mail-type=ALL\n",
    "\n",
    "module load openmpi/intel/3.1.1\n",
    "module load beagle-lib/2.1.2\n",
    "\n",
    "python /home2/s185491/p/xiaolongTools/project/Junonia/runMrBayes_largeInput.py -i /work/biophysics/s185491/20181226Calephelis/Phylo/mito/Calephelis266.mito\n",
    "```\n",
    "\n",
    "```\n",
    "cp /work/biophysics/s185491/20181226Calephelis/Phylo/mitoNewick/Calephelis266.mito /work/biophysics/s185491/20181226Calephelis/Phylo/mito/Calephelis266.mito.newick\n",
    " cp /work/biophysics/s185491/20181226Calephelis/Phylo/mitoMrBayes/Calephelis266.mito /work/biophysics/s185491/20181226Calephelis/Phylo/mito/Calephelis266.mito.nex\n",
    "#rename\n",
    "/home2/mtang/anaconda2/bin/python /work/biophysics/mtang/SNP_calling/scripts/rename_tree_fullname.py /work/biophysics/s185491/20181226Calephelis/Phylo/mito/Calephelis266.mito.newick\n",
    "```\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### make whole genome, CDS, intron tree"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### extract sequences\n",
    "```\n",
    "import sys\n",
    "sys.path.append('/home/xcao/p/xiaolongTools/utils/mapFile')\n",
    "import mapfileIO\n",
    "import glob\n",
    "\n",
    "folder = '/home/xcao/w/2018Calephelis/summary/map.npInt8.0/'\n",
    "outfile = '/home/xcao/w/2018Calephelis/20181226Phylo/whole/Calephelis266.wholeGap60'\n",
    "file_filter = '/home/xcao/w/2018Calephelis/summary/map.sitesInfo/gapCount266'\n",
    "files = glob.glob(folder+'*')\n",
    "print('total map npInt8 files', len(files))\n",
    "\n",
    "mapfileIO.npInt8stoFastas(files, threads=32, output=outfile, headerFun=lambda x:x.split('_')[0], npFilter=file_filter, npFilterFun=lambda x:x<=60, merged=True)\n",
    "\n",
    "outfile = '/home/xcao/w/2018Calephelis/20181226Phylo/whole/Calephelis266.wholeGap13'\n",
    "mapfileIO.npInt8stoFastas(files, threads=32, output=outfile, headerFun=lambda x:x.split('_')[0], npFilter=file_filter, npFilterFun=lambda x:x<=13, merged=True)\n",
    "\n",
    "file_filter = '/home/xcao/w/2018Calephelis/summary/map.sitesInfo/gff.0inter.2intron.6CDS'\n",
    "outfile = '/home/xcao/w/2018Calephelis/20181226Phylo/CDS/Calephelis266.CDS'\n",
    "mapfileIO.npInt8stoFastas(files, threads=32, output=outfile, headerFun=lambda x:x.split('_')[0], npFilter=file_filter, npFilterFun=lambda x:x==6, merged=True)\n",
    "\n",
    "file_filter = '/home/xcao/w/2018Calephelis/summary/map.sitesInfo/gff.0inter.2intron.6CDS'\n",
    "outfile = '/home/xcao/w/2018Calephelis/20181226Phylo/intron/Calephelis266.intron'\n",
    "mapfileIO.npInt8stoFastas(files, threads=32, output=outfile, headerFun=lambda x:x.split('_')[0], npFilter=file_filter, npFilterFun=lambda x:x==2, merged=True)\n",
    "\n",
    "file_filter1 = '/home/xcao/w/2018Calephelis/summary/map.sitesInfo/gff.0inter.2intron.6CDS'\n",
    "file_filter2 = '/home/xcao/w/2018Calephelis/summary/map.sitesInfo/gapCount266'\n",
    "npFilter1 = mapfileIO.loadMapBinary(file_filter1)\n",
    "npFilter2 = mapfileIO.loadMapBinary(file_filter2)\n",
    "\n",
    "outfile = '/home/xcao/w/2018Calephelis/20181226Phylo/CDS/Calephelis266.CDS.Gap26'\n",
    "mapfileIO.npInt8stoFastas(files, threads=32, output=outfile, headerFun=lambda x:x.split('_')[0], npFilter=((npFilter1==6) & (npFilter2<=26)), npFilterFun=None, merged=True)\n",
    "\n",
    "outfile = '/home/xcao/w/2018Calephelis/20181226Phylo/intron/Calephelis266.intron.Gap26'\n",
    "mapfileIO.npInt8stoFastas(files, threads=32, output=outfile, headerFun=lambda x:x.split('_')[0], npFilter=((npFilter1==2) & (npFilter2<=26)), npFilterFun=None, merged=True)\n",
    "\n",
    "outfile = '/home/xcao/w/2018Calephelis/20181226Phylo/intron/Calephelis266.intron.Gap100'\n",
    "mapfileIO.npInt8stoFastas(files, threads=32, output=outfile, headerFun=lambda x:x.split('_')[0], npFilter=((npFilter1==2) & (npFilter2<=100)), npFilterFun=None, merged=True)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### transfer files\n",
    "```\n",
    "scp -r /home/xcao/w/2018Calephelis/20181226Phylo/whole/ s185491@nucleus.biohpc.swmed.edu:/work/biophysics/s185491/20181226Calephelis/Phylo/\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### run examl\n",
    "\n",
    "run examl for the small file to generate the starting tree. Then split the CDS, whole genome and intron sequences to 5 parts, and run ExaML\n",
    "```\n",
    "python /home2/s185491/p/xiaolongTools/utils/splitFastaAlignments2NPartsContinuingEqualStepOrRandom.py -t 48 -m 1 -N 5 -i /work/biophysics/s185491/20181226Calephelis/Phylo/CDS/Calephelis266.CDS -o /work/biophysics/s185491/20181226Calephelis/Phylo/CDS/\n",
    "\n",
    "python /home2/s185491/p/xiaolongTools/utils/splitFastaAlignments2NPartsContinuingEqualStepOrRandom.py -t 48 -m 1 -N 5 -i /work/biophysics/s185491/20181226Calephelis/Phylo/intron/Calephelis266.intron.Gap100 -o /work/biophysics/s185491/20181226Calephelis/Phylo/intron/\n",
    "\n",
    "python /home2/s185491/p/xiaolongTools/utils/splitFastaAlignments2NPartsContinuingEqualStepOrRandom.py -t 48 -m 1 -N 5 -i /work/biophysics/s185491/20181226Calephelis/Phylo/whole/Calephelis266.wholeGap60 -o /work/biophysics/s185491/20181226Calephelis/Phylo/whole/\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "txt_qsub = '''#!/bin/bash\n",
    "#SBATCH --job-name={key}_examl\n",
    "#SBATCH --partition=32GB\n",
    "#SBATCH --nodes=10\n",
    "#SBATCH --ntasks=320\n",
    "#SBATCH --time=5-00:00:00\n",
    "#SBATCH --output=0.%j.out\n",
    "#SBATCH --mail-user=xiaolong.cao@utsouthwestern.edu\n",
    "#SBATCH --mail-type=ALL\n",
    "\n",
    "module load openmpi/intel/3.1.1\n",
    "\n",
    "\n",
    "'''\n",
    "commandline = '''/home2/s185491/p/anaconda3/anaconda520/bin/python3 /home2/s185491/p/xiaolongTools/utils/runExaML_withFastaForLargeFiles.py -i {filename} -T 320 -r {file_ref}\n",
    "'''\n",
    "\n",
    "files = '''/work/biophysics/s185491/20181226Calephelis/Phylo/zChr/Calephelis266.zChr.split5.0\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/zChr/Calephelis266.zChr.split5.1\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/zChr/Calephelis266.zChr.split5.2\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/zChr/Calephelis266.zChr.split5.3\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/zChr/Calephelis266.zChr.split5.4\n",
    "'''.split()\n",
    "\n",
    "dc_files = {}\n",
    "dc_ref = {}\n",
    "dc_files['CDS'] = '/work/biophysics/s185491/20181226Calephelis/Phylo/CDS/Calephelis266.CDS.split5.{N}'\n",
    "dc_ref['CDS'] = '/work/biophysics/s185491/20181226Calephelis/Phylo/CDS/ExaML_result.Calephelis266.CDS.Gap26'\n",
    "dc_files['intron'] = '/work/biophysics/s185491/20181226Calephelis/Phylo/intron/Calephelis266.intron.Gap100.split5.{N}'\n",
    "dc_ref['intron'] = '/work/biophysics/s185491/20181226Calephelis/Phylo/intron/ExaML_result.Calephelis266.intron.Gap26'\n",
    "dc_files['whole'] = '/work/biophysics/s185491/20181226Calephelis/Phylo/whole/Calephelis266.wholeGap60.split5.{N}'\n",
    "dc_ref['whole'] = '/work/biophysics/s185491/20181226Calephelis/Phylo/whole/ExaML_result.Calephelis266.wholeGap13'\n",
    "\n",
    "for k in dc_files:\n",
    "    ls_files = [dc_files[k].format(N = i) for i in range(5)]\n",
    "    file_ref = dc_ref[k]\n",
    "    fout = open('/work/biophysics/s185491/20181226Calephelis/Phylo/cmds/qsub_examl_'+k,'w')\n",
    "    fout.write(txt_qsub.format(key=k))\n",
    "    for filename in ls_files:\n",
    "        fout.write(commandline.format(filename=filename, file_ref=file_ref))\n",
    "    fout.close()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### combine whole genome trees\n",
    "```\n",
    " cat /work/biophysics/s185491/20181226Calephelis/Phylo/whole/ExaML_result.Calephelis266.wholeGap60.split5.* >/work/biophysics/s185491/20181226Calephelis/Phylo/whole/ExaML_result.Calephelis266.wholeGap60.all5\n",
    " /home2/s185491/p/anaconda3/anaconda520/bin/sumtrees.py -f 0.00 /work/biophysics/s185491/20181226Calephelis/Phylo/whole/ExaML_result.Calephelis266.wholeGap60.all5 >/work/biophysics/s185491/20181226Calephelis/Phylo/whole/ExaML_result.Calephelis266.wholeGap60.all5.sum\n",
    " python /home2/s185491/p/xiaolongTools/utils/trees/sumtreesNEXUS2newick.py -i /work/biophysics/s185491/20181226Calephelis/Phylo/whole/ExaML_result.Calephelis266.wholeGap60.all5.sum\n",
    " /home2/mtang/anaconda2/bin/python /work/biophysics/mtang/SNP_calling/scripts/rename_tree_fullname.py /work/biophysics/s185491/20181226Calephelis/Phylo/whole/ExaML_result.Calephelis266.wholeGap60.all5.sum.nwk\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### combine zChr\n",
    "```\n",
    " cat /work/biophysics/s185491/20181226Calephelis/Phylo/whole/ExaML_result.Calephelis266.wholeGap60.split5.* >/work/biophysics/s185491/20181226Calephelis/Phylo/whole/ExaML_result.Calephelis266.wholeGap60.all5\n",
    " /home2/s185491/p/anaconda3/anaconda520/bin/sumtrees.py -f 0.00 /work/biophysics/s185491/20181226Calephelis/Phylo/whole/ExaML_result.Calephelis266.wholeGap60.all5 >/work/biophysics/s185491/20181226Calephelis/Phylo/whole/ExaML_result.Calephelis266.wholeGap60.all5.sum\n",
    " python /home2/s185491/p/xiaolongTools/utils/trees/sumtreesNEXUS2newick.py -i /work/biophysics/s185491/20181226Calephelis/Phylo/whole/ExaML_result.Calephelis266.wholeGap60.all5.sum\n",
    " /home2/mtang/anaconda2/bin/python /work/biophysics/mtang/SNP_calling/scripts/rename_tree_fullname.py /work/biophysics/s185491/20181226Calephelis/Phylo/whole/ExaML_result.Calephelis266.wholeGap60.all5.sum.nwk\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### combine intron\n",
    "```\n",
    " cat /work/biophysics/s185491/20181226Calephelis/Phylo/intron/ExaML_result.Calephelis266.intron.Gap100.split5.* >/work/biophysics/s185491/20181226Calephelis//Phylo/intron/ExaML_result.Calephelis266.intron.Gap100.all5\n",
    " /home2/s185491/p/anaconda3/anaconda520/bin/sumtrees.py -f 0.00 /work/biophysics/s185491/20181226Calephelis//Phylo/intron/ExaML_result.Calephelis266.intron.Gap100.all5 >/work/biophysics/s185491/20181226Calephelis//Phylo/intron/ExaML_result.Calephelis266.intron.Gap100.all5.sum\n",
    " python /home2/s185491/p/xiaolongTools/utils/trees/sumtreesNEXUS2newick.py -i /work/biophysics/s185491/20181226Calephelis//Phylo/intron/ExaML_result.Calephelis266.intron.Gap100.all5.sum\n",
    " /home2/mtang/anaconda2/bin/python /work/biophysics/mtang/SNP_calling/scripts/rename_tree_fullname.py /work/biophysics/s185491/20181226Calephelis//Phylo/intron/ExaML_result.Calephelis266.intron.Gap100.all5.sum.nwk\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### combine CDS\n",
    "```\n",
    " cat /work/biophysics/s185491/20181226Calephelis/Phylo/CDS/ExaML_result.Calephelis266.CDS.split5.* >/work/biophysics/s185491/20181226Calephelis/Phylo/CDS/ExaML_result.Calephelis266.CDS.all5\n",
    " /home2/s185491/p/anaconda3/anaconda520/bin/sumtrees.py -f 0.00 /work/biophysics/s185491/20181226Calephelis/Phylo/CDS/ExaML_result.Calephelis266.CDS.all5 >/work/biophysics/s185491/20181226Calephelis/Phylo/CDS/ExaML_result.Calephelis266.CDS.all5.sum\n",
    " python /home2/s185491/p/xiaolongTools/utils/trees/sumtreesNEXUS2newick.py -i /work/biophysics/s185491/20181226Calephelis/Phylo/CDS/ExaML_result.Calephelis266.CDS.all5.sum\n",
    " /home2/mtang/anaconda2/bin/python /work/biophysics/mtang/SNP_calling/scripts/rename_tree_fullname.py /work/biophysics/s185491/20181226Calephelis/Phylo/CDS/ExaML_result.Calephelis266.CDS.all5.sum.nwk\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Phylogenetic Tree for C. rawsoni group"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### count gaps\n",
    "\n",
    "gap ratio 45.69%"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "import sys\n",
    "sys.path.append('/home/xcao/p/xiaolongTools/utils/mapFile')\n",
    "import mapfileIO\n",
    "import glob\n",
    "import numpy as np\n",
    "import pickle\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "folderBioHPC = '/work/archive/biophysics/Nick_lab/s185491/Celastrina/fq'\n",
    "folder_fq = '/archive/butterfly/ready_fastq/'\n",
    "folder_summary = '/home/xcao/w/2018Calephelis/summary/'\n",
    "\n",
    "file_summary = os.path.join(folder_summary,'20181217Calephelis.xlsx')\n",
    "df_summary = pd.read_excel(file_summary,sheet_name='Sheet1')\n",
    "df_summary = df_summary[df_summary[\"Filter_rawsoni\"]==1]\n",
    "sample_prefix = [e[1:] for e in df_summary['ID']]\n",
    "print('unique sample prefix', len(set(sample_prefix)))\n",
    "\n",
    "folder = '/home/xcao/w/2018Calephelis/summary/map.npInt8.0/'\n",
    "outfile = '/home/xcao/w/2018Calephelis/summary/map.sitesInfo/gapCount.rawsoni35'\n",
    "outfileHist = outfile + '.hist'\n",
    "files = glob.glob(folder+'*')\n",
    "files = [f for f in files if os.path.basename(f).split('_')[0] in sample_prefix]\n",
    "\n",
    "print('total map files', len(files))\n",
    "\n",
    "#ls_mapInt8 = mapfileIO.loadMapBinaries(files, threads=8)#multithreading not working. use the slow method\n",
    "ls_mapInt8 = [mapfileIO.loadMapBinary(f) for f in files]\n",
    "temp = ls_mapInt8[0]\n",
    "maplen = len(temp)\n",
    "gapCount = np.zeros(maplen,dtype=np.int16)\n",
    "for e in ls_mapInt8:\n",
    "    gapCount += (e==45)\n",
    "\n",
    "with open(outfile,'wb') as f:\n",
    "    pickle.dump(gapCount,f)\n",
    "\n",
    "gapHist = np.unique(gapCount,return_counts=True)\n",
    "with open(outfileHist,'w') as f:\n",
    "    for c, p in zip(*gapHist):\n",
    "        f.write('{}\\t{}\\n'.format(c,p))\n",
    "\n",
    "#average gap per site\n",
    "gapAvg = gapCount.sum() / maplen\n",
    "print('average gap per site', gapAvg)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### extract sequences of mito, Z, CDS, whole and intron"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "previous step continued\n",
    "```\n",
    "npFilter_gap = gapCount\n",
    "npFilter_sites = mapfileIO.loadMapBinary(filename='/home/xcao/w/2018Calephelis/summary/map.sitesInfo/gff.0inter.2intron.6CDS')\n",
    "dc_files = {os.path.basename(k).split('_')[0]:v for k,v in zip(files, ls_mapInt8)}\n",
    "\n",
    "outfolder = '/home/xcao/w/2018Calephelis/20190101rawsoni/'\n",
    "outfile = os.path.join(outfolder,'Calephelis35.wholeGap7')\n",
    "mapfileIO.npInt8stoFastas(dc_files, threads=32, output=outfile, headerFun=lambda x:x.split('_')[0], npFilter=npFilter_gap, npFilterFun=lambda x:x<=7, merged=True)\n",
    "\n",
    "outfile = os.path.join(outfolder,'Calephelis35.CDS')\n",
    "mapfileIO.npInt8stoFastas(dc_files, threads=32, output=outfile, headerFun=lambda x:x.split('_')[0], npFilter=npFilter_sites, npFilterFun=lambda x:x==6, merged=True)\n",
    "\n",
    "\n",
    "def extractSeq(file_in, file_out, keepIDs):\n",
    "    '''\n",
    "    extract sequence from file_in and write to file_out if seq.id in keepIDs\n",
    "    '''\n",
    "    count_filein = 0\n",
    "    count_fileout = 0\n",
    "    keepIDs = set(keepIDs)\n",
    "    fout = open(file_out,'w')\n",
    "    for s in SeqIO.parse(file_in,'fasta'):\n",
    "        count_filein += 1\n",
    "        if s.id in keepIDs:\n",
    "            count_fileout += 1\n",
    "            fout.write('>'+s.id+'\\n'+str(s.seq)+'\\n')\n",
    "    fout.close()\n",
    "    print('input seqnum:', count_filein,'output seqnum:', count_fileout)\n",
    "    return None\n",
    "\n",
    "outfile = os.path.join(outfolder,'Calephelis35.zChr')\n",
    "file_in = '/home/xcao/w/2018Calephelis/20181226Phylo/zChrBased3Ref/Calephelis266.zChr'\n",
    "extractSeq(file_in,outfile, sample_prefix)\n",
    "\n",
    "outfile = os.path.join(outfolder,'Calephelis35.mito')\n",
    "file_in = '/home/xcao/w/2018Calephelis/20181226Phylo/mitochondria/Calephelis266.mito'\n",
    "extractSeq(file_in,outfile, sample_prefix)\n",
    "\n",
    "outfile = os.path.join(outfolder,'Calephelis35.intron')\n",
    "file_in = '/home/xcao/w/2018Calephelis/20181226Phylo/intron/Calephelis266.intron'\n",
    "extractSeq(file_in,outfile, sample_prefix)\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### transfer files and run tree"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "transfer\n",
    "```\n",
    "scp -r /home/xcao/w/2018Calephelis/20190101rawsoni/ s185491@nucleus.biohpc.swmed.edu:/work/biophysics/s185491/20181226Calephelis/Phylo\n",
    "```\n",
    "\n",
    "run mitochondria\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "mito is simple. Change some words.\n",
    "```\n",
    "python /home2/s185491/p/xiaolongTools/utils/splitFastaAlignments2NPartsContinuingEqualStepOrRandom.py -t 48 -m 1 -N 5 -i /work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/Calephelis35.CDS -o /work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/\n",
    "\n",
    "python /home2/s185491/p/xiaolongTools/utils/splitFastaAlignments2NPartsContinuingEqualStepOrRandom.py -t 48 -m 1 -N 5 -i /work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/Calephelis35.intron -o /work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/\n",
    "\n",
    "python /home2/s185491/p/xiaolongTools/utils/splitFastaAlignments2NPartsContinuingEqualStepOrRandom.py -t 48 -m 1 -N 5 -i /work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/Calephelis35.wholeGap7 -o /work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/\n",
    "\n",
    "python /home2/s185491/p/xiaolongTools/utils/splitFastaAlignments2NPartsContinuingEqualStepOrRandom.py -t 48 -m 1 -N 5 -i /work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/Calephelis35.zChr -o /work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/\n",
    "\n",
    "```\n",
    "\n",
    "generate running scripts\n",
    "```\n",
    "txt_qsub = '''#!/bin/bash\n",
    "#SBATCH --job-name=examl\n",
    "#SBATCH --partition=32GB\n",
    "#SBATCH --nodes=10\n",
    "#SBATCH --ntasks=320\n",
    "#SBATCH --time=5-00:00:00\n",
    "#SBATCH --output=0.%j.out\n",
    "#SBATCH --mail-user=xiaolong.cao@utsouthwestern.edu\n",
    "#SBATCH --mail-type=ALL\n",
    "\n",
    "module load openmpi/intel/3.1.1\n",
    "\n",
    "\n",
    "'''\n",
    "commandline = '''/home2/s185491/p/anaconda3/anaconda520/bin/python3 /home2/s185491/p/xiaolongTools/utils/runExaML_withFastaForLargeFiles.py -i {filename} -T 320 -r {file_ref}\n",
    "'''\n",
    "\n",
    "\n",
    "dc_files = {}\n",
    "dc_ref = {}\n",
    "dc_files['CDS'] = '/work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/Calephelis35.CDS.split5.{N}'\n",
    "dc_files['intron'] = '/work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/Calephelis35.intron.split5.{N}'\n",
    "dc_files['whole'] = '/work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/Calephelis35.wholeGap7.split5.{N}'\n",
    "dc_files['zChr'] = '/work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/Calephelis35.zChr.split5.{N}'\n",
    "\n",
    "for k in dc_files:\n",
    "    ls_files = [dc_files[k].format(N = i) for i in range(5)]\n",
    "    file_ref = '/work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/Calephelis35.mito'\n",
    "    fout = open('/work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/cmds/qsub_examl_'+k,'w')\n",
    "    fout.write(txt_qsub.format(key=k))\n",
    "    for filename in ls_files:\n",
    "        fout.write(commandline.format(filename=filename, file_ref=file_ref))\n",
    "    fout.close()\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " ```\n",
    "txt = ''' cat {keyword}.split5.* >{keyword}.all5\n",
    " /home2/s185491/p/anaconda3/anaconda520/bin/sumtrees.py -f 0.00 {keyword}.all5 >{keyword}.all5.sum\n",
    " python /home2/s185491/p/xiaolongTools/utils/trees/sumtreesNEXUS2newick.py -i {keyword}.all5.sum\n",
    " /home2/mtang/anaconda2/bin/python /work/biophysics/mtang/SNP_calling/scripts/rename_tree_fullname.py {keyword}.all5.sum.nwk\n",
    "'''\n",
    "\n",
    "keywords = '''/work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/ExaML_result.Calephelis35.CDS\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/ExaML_result.Calephelis35.intron\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/ExaML_result.Calephelis35.wholeGap7\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/ExaML_result.Calephelis35.zChr\n",
    "'''.split()\n",
    "\n",
    "for keyword in keywords:\n",
    "    print(txt.format(keyword=keyword))\n",
    "\n",
    " ```\n",
    "```\n",
    "/home2/mtang/anaconda2/bin/python /work/biophysics/mtang/SNP_calling/scripts/rename_tree_fullname.py /work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/Calephelis35.mito\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### zCDS tree"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "get sequence\n",
    "```\n",
    "import sys\n",
    "sys.path.append('/home/xcao/p/xiaolongTools/utils/mapFile')\n",
    "import mapfileIO\n",
    "import glob\n",
    "import numpy as np\n",
    "import pickle\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "folderBioHPC = '/work/archive/biophysics/Nick_lab/s185491/Celastrina/fq'\n",
    "folder_fq = '/archive/butterfly/ready_fastq/'\n",
    "folder_summary = '/home/xcao/w/2018Calephelis/summary/'\n",
    "\n",
    "file_summary = os.path.join(folder_summary,'20181217Calephelis.xlsx')\n",
    "df_summary = pd.read_excel(file_summary,sheet_name='Sheet1')\n",
    "df_summary = df_summary[df_summary[\"Filter_rawsoni\"]==1]\n",
    "sample_prefix = [e[1:] for e in df_summary['ID']]\n",
    "print('unique sample prefix', len(set(sample_prefix)))\n",
    "\n",
    "folder = '/home/xcao/w/2018Calephelis/summary/map.npInt8.0/'\n",
    "outfile = '/home/xcao/w/2018Calephelis/summary/map.sitesInfo/gapCount.rawsoni35'\n",
    "outfileHist = outfile + '.hist'\n",
    "files = glob.glob(folder+'*')\n",
    "files = [f for f in files if os.path.basename(f).split('_')[0] in sample_prefix]\n",
    "\n",
    "print('total map files', len(files))\n",
    "\n",
    "#ls_mapInt8 = mapfileIO.loadMapBinaries(files, threads=8)#multithreading not working. use the slow method\n",
    "ls_mapInt8 = [mapfileIO.loadMapBinary(f) for f in files]\n",
    "dc_mapInt8 = {}\n",
    "for k,v in zip(files,ls_mapInt8):\n",
    "    k = os.path.basename(k).split('_')[0]\n",
    "    dc_mapInt8[k] = v\n",
    "\n",
    "npFilter_z = mapfileIO.loadMapBinary(filename='/home/xcao/w/2018Calephelis/summary/map.sitesInfo/zChrFilter')\n",
    "npFilter_sites = mapfileIO.loadMapBinary(filename='/home/xcao/w/2018Calephelis/summary/map.sitesInfo/gff.0inter.2intron.6CDS')\n",
    "npFilter_z.sum() / npFilter_z.shape[0]\n",
    "npFilter_zCDS = (npFilter_z) & (npFilter_sites == 6)\n",
    "\n",
    "with open('/home/xcao/w/2018Calephelis/summary/map.sitesInfo/zChrCDSFilter','wb') as f:\n",
    "    pickle.dump(npFilter_zCDS, f)\n",
    "\n",
    "outfile = '/home/xcao/w/2018Calephelis/20190101rawsoni/Calephelis35.zChrCDS'\n",
    "mapfileIO.npInt8stoFastas(dc_mapInt8, threads=32, output=outfile, headerFun=lambda x:x.split('_')[0], npFilter=npFilter_zCDS, npFilterFun=lambda x:x<=7, merged=True)\n",
    "\n",
    "#mapfileIO.npInt8stoFastas(files,threads=32, output= '/home/xcao/w/2018Calephelis/20190101rawsoni/Calephelis35.zChrCDS', headerFun=lambda x:x.split('_')[0], npFilter='/home/xcao/w/2018Calephelis/summary/map.sitesInfo/zChrCDSFilter',merged=True)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "run RAxML\n",
    "```\n",
    "scp /home/xcao/w/2018Calephelis/20190101rawsoni/Calephelis35.zChrCDS /work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "/home2/mtang/anaconda2/bin/python /work/biophysics/mtang/SNP_calling/scripts/rename_tree_fullname.py /work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/RAxML_bipartitions.Calephelis35.zChrCDS\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### gene tree simple 35 C. rawsoni"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### get gene trees by simply dividing the sequences to 50kb window\n",
    "\n",
    "```\n",
    "python /home2/s185491/p/xiaolongTools/utils/splitFastaAlignments2NPartsContinuingEqualStepOrRandom.py -i /work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/Calephelis35.wholeGap7 -o /work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/wholeGap7/ -b 25000 -m 0 -t 48 -w 50000\n",
    "\n",
    "python /home2/s185491/p/xiaolongTools/utils/splitFastaAlignments2NPartsContinuingEqualStepOrRandom.py -i /work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/Calephelis35.CDS -o /work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/CDS/ -b 25000 -m 0 -t 48 -w 50000\n",
    "\n",
    "python /home2/s185491/p/xiaolongTools/utils/splitFastaAlignments2NPartsContinuingEqualStepOrRandom.py -i /work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/Calephelis35.intron -o /work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/intron/ -b 25000 -m 0 -t 48 -w 50000\n",
    "\n",
    "python /home2/s185491/p/xiaolongTools/utils/splitFastaAlignments2NPartsContinuingEqualStepOrRandom.py -i /work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/Calephelis35.zChr -o /work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/zChr/ -b 25000 -m 0 -t 48 -w 50000\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### generate running script\n",
    "```\n",
    "folders = '''/work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/CDS/\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/intron/\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/wholeGap7/\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/zChr/\n",
    "'''.split()\n",
    "folder_cmds = '/work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/cmds/'\n",
    "\n",
    "import glob\n",
    "import os\n",
    "\n",
    "files = []\n",
    "for f in folders:\n",
    "    files += glob.glob(f+'*')\n",
    "print(len(files))\n",
    "\n",
    "qusb_txt='''#!/bin/bash\n",
    "#SBATCH --job-name=raxml\n",
    "#SBATCH --partition=32GB\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=32\n",
    "#SBATCH --time=5-00:00:00\n",
    "#SBATCH --output=0.%j.out\n",
    "\n",
    "python /home2/s185491/p/xiaolongTools/utils/runRAxMLForIndividualSmallFiles.py -i {file_in} -T 32 -N 50\n",
    "\n",
    "'''\n",
    "for f in files:\n",
    "    open(os.path.join(folder_cmds,os.path.basename(f)), 'w').write(qusb_txt.format(file_in=f))\n",
    "\n",
    "file_cmds = glob.glob(folder_cmds+'*')\n",
    "open(folder_cmds[:-1]+'.all','w').write('\\n'.join(file_cmds))\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "submit the jobs\n",
    "```\n",
    " nohup python /home2/s185491/p/xiaolongTools/utils/submitJobs.py -i /work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/cmds.all -N 32 -t 20 -k raxml &\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### sumarize trees"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "combine outputs\n",
    "```\n",
    "cat /work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/CDS.RAxMLbestTree/* >/work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/CDS.RAxMLbestTree.all\n",
    "cat /work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/intron.RAxMLbestTree/* >/work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/intron.RAxMLbestTree.all\n",
    "cat /work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/wholeGap7.RAxMLbestTree/* >/work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/wholeGap7.RAxMLbestTree.all\n",
    "cat /work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/zChr.RAxMLbestTree/* >/work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/zChr.RAxMLbestTree.all\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### sumtrees\n",
    "```\n",
    "txt = '''\n",
    "/home2/s185491/p/anaconda3/anaconda520/bin/sumtrees.py --ignore-missing-support -f 0.00 {keyword} >{keyword}.sum\n",
    "python /home2/s185491/p/xiaolongTools/utils/trees/sumtreesNEXUS2newick.py -i {keyword}.sum\n",
    "/home2/mtang/anaconda2/bin/python /work/biophysics/mtang/SNP_calling/scripts/rename_tree_fullname.py {keyword}.sum.nwk\n",
    "'''\n",
    "\n",
    "keywords = '''/work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/CDS.RAxMLbestTree.all\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/intron.RAxMLbestTree.all\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/wholeGap7.RAxMLbestTree.all\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/zChr.RAxMLbestTree.all\n",
    "\n",
    "'''.split()\n",
    "for keyword in keywords:\n",
    "    print(txt.format(keyword=keyword))\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Astral\n",
    "```\n",
    "\n",
    "txt = '''java -Xmx3000m -jar /home2/s185491/p/Astral/astral.5.6.2.jar -i {f} -o {f}.astral '''\n",
    "\n",
    "files = '''/work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/CDS.RAxMLbestTree.all\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/intron.RAxMLbestTree.all\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/wholeGap7.RAxMLbestTree.all\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/zChr.RAxMLbestTree.all'''.split()\n",
    "\n",
    "for f in files:\n",
    "    print(txt.format(f=f))\n",
    "\n",
    "files = '''/work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/CDS.RAxMLbestTree.all.astral\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/intron.RAxMLbestTree.all.astral\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/wholeGap7.RAxMLbestTree.all.astral\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/zChr.RAxMLbestTree.all.astral\n",
    "'''.split()\n",
    "for f in files:\n",
    "    print('/home2/mtang/anaconda2/bin/python /work/biophysics/mtang/SNP_calling/scripts/rename_tree_fullname.py  ' + f)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### densiTree\n",
    "```\n",
    "\n",
    "filename = r\"C:\\P\\DensiTree\\CDS.RAxMLbestTree.all.rescale2\"\n",
    "filename = r\"C:\\P\\DensiTree\\wholeGap7.RAxMLbestTree.all.rescale2\"\n",
    "\n",
    "from ete3 import Tree\n",
    "lines = open(filename).readlines()\n",
    "ls_trees = [Tree(e.strip()) for e in lines]\n",
    "ls_trees.sort(key=lambda x:len(x.get_leaf_names()), reverse=True)\n",
    "fout = open(filename,'w')\n",
    "for t in ls_trees:\n",
    "    fout.write(t.write()+'\\n')\n",
    "fout.close()\n",
    "\n",
    "file_talbe = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\UTSW\\2018Calephelis\\phylo\\35rawsoni_gene_trees\\densitree\\nameTable.txt\"\n",
    "\n",
    "dc_name = {}\n",
    "for line in open(file_talbe):\n",
    "    es = line.strip().replace(' ','_').split('\\t')\n",
    "    dc_name[es[0]] = '_'.join(es[1:])\n",
    "    \n",
    "filename = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\UTSW\\2018Calephelis\\phylo\\35rawsoni_gene_trees\\densitree\\CDS.svg\"\n",
    "filename = r\"C:\\Users\\ATPs\\OneDrive\\Lab\\UTSW\\2018Calephelis\\phylo\\35rawsoni_gene_trees\\densitree\\whole.svg\"\n",
    "txt = open(filename).read()\n",
    "for k,v in dc_name.items():\n",
    "    txt = txt.replace(k+'</text>\\n', v+'</text>\\n')\n",
    "open(filename,'w').write(txt)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### gene trees find subtrees that group rawsoni, freemani, arizonensis together"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "transfer trees from bioHPC to alea\n",
    "```\n",
    "scp -r s185491@nucleus.biohpc.swmed.edu:/work/biophysics/s185491/20181226Calephelis/Phylo/20190103rawsoniGeneTree/wholeGap7.RAxMLbestTree/  /home/xcao/w/2018Calephelis/20190101rawsoni/wholeGenomeGeneTrees\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "get the trees\n",
    "```\n",
    "from ete3 import Tree\n",
    "from multiprocessing import Pool\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "#filename = r\"C:\\Users\\K\\Downloads\\chr10_871\"\n",
    "#tree = Tree(open(filename).read())\n",
    "\n",
    "def addNodeNames(tree):\n",
    "    '''\n",
    "    the internal nodes have no names. add unique names like node1, node2 to them\n",
    "    '''\n",
    "    n = 1\n",
    "    for node in tree.traverse():\n",
    "        if not node.is_leaf():\n",
    "            node.name = 'node' + str(n)\n",
    "            n += 1\n",
    "    return None\n",
    "\n",
    "def add_species(tree):\n",
    "    '''\n",
    "    add speces based on name for leaves\n",
    "    '''\n",
    "    for leaf in tree.iter_leaves():\n",
    "        if leaf.name in dc_sample2species:\n",
    "            leaf.add_features(species=dc_sample2species[leaf.name])\n",
    "    return None\n",
    "\n",
    "def get_leaf_species(tree):\n",
    "    '''\n",
    "    get the species names for a tree node\n",
    "    '''\n",
    "    return([e.species for e in tree.get_leaves()])\n",
    "\n",
    "def checkIfSpeciesMono(tree, species):\n",
    "    '''\n",
    "    given a tree, leaf belong to species form a monotree\n",
    "    '''\n",
    "    tree = tree.copy()\n",
    "    tree.unroot()\n",
    "    leaves_test = []\n",
    "    for leaf in tree.iter_leaves():\n",
    "        if leaf.species == species:\n",
    "            leaves_test.append(leaf.name)\n",
    "    \n",
    "    for node in tree.iter_descendants():\n",
    "        tree2 = tree.copy()\n",
    "        tree2.set_outgroup(node.name)\n",
    "        test_root = tree.get_common_ancestor(*leaves_test)\n",
    "        species_identified = get_leaf_species(test_root)\n",
    "        species_identified = [e for e in species_identified if '?' not in e] #ignore species with '?'\n",
    "        if len(species_identified) == len(leaves_test):\n",
    "            return True\n",
    "    return False\n",
    "\n",
    "def processTXT(txt):\n",
    "    tree_id, tree_txt = txt.split()\n",
    "    tree = Tree(tree_txt)\n",
    "    addNodeNames(tree)\n",
    "    add_species(tree)\n",
    "    tree.unroot()\n",
    "    addNodeNames(tree)\n",
    "    dc ={}\n",
    "    dc['tree_id']=tree_id\n",
    "    dc['tree'] = tree\n",
    "    dc['tree_txt'] = tree_txt\n",
    "    species_checking = ['Calephelis arizonensis','Calephelis freemani','Calephelis nuevoleon','Calephelis rawsoni']\n",
    "    for s in species_checking:\n",
    "        dc[s] = checkIfSpeciesMono(tree, s)\n",
    "    return dc\n",
    "\n",
    "\n",
    "# prepare samples and species names\n",
    "file_sample2species = '/home/xcao/w/2018Calephelis/20190101rawsoni/wholeGenomeGeneTrees/id2species'\n",
    "\n",
    "dc_sample2species = {}\n",
    "for line in open(file_sample2species):\n",
    "    line = line.strip()\n",
    "    k,v = line.split('\\t')\n",
    "    dc_sample2species[k] = v\n",
    "\n",
    "#remove outgroup in dc_sample2species\n",
    "testing_groups = ['arizonensis', 'freemani','nuevoleon','rawsoni']\n",
    "samples_keep = []\n",
    "for k,v in dc_sample2species.items():\n",
    "    keep = False\n",
    "    for e in testing_groups:\n",
    "        if e in v:\n",
    "            keep = True\n",
    "            break\n",
    "    if keep:\n",
    "        samples_keep.append(k)\n",
    "print(len(samples_keep),'samples keep in testing_groups')\n",
    "dc_sample2species = {k:v for k,v in dc_sample2species.items() if k in samples_keep}\n",
    "\n",
    "#prepare trees\n",
    "folder = '/home/xcao/w/2018Calephelis/20190101rawsoni/wholeGenomeGeneTrees/wholeGap7.RAxMLbestTree/'\n",
    "import glob\n",
    "files = glob.glob(folder+'*')\n",
    "print('files of gene trees', len(files))\n",
    "files.sort(key=lambda x:int(x.split('.')[-1]))\n",
    "#combine the trees\n",
    "fout = open(folder.rstrip('/')+'.all', 'w')\n",
    "for f in files:\n",
    "    n = f.split('.')[-1]\n",
    "    t = open(f).read().strip()\n",
    "    fout.write(n+'\\t'+t+'\\n')\n",
    "fout.close()\n",
    "\n",
    "#remove \n",
    "filename = '/home/xcao/w/2018Calephelis/20190101rawsoni/wholeGenomeGeneTrees/wholeGap7.RAxMLbestTree.all'\n",
    "fileout = filename+'.26'\n",
    "fout = open(fileout,'w')\n",
    "\n",
    "def removeNodes(t,dc_sample2species):\n",
    "    '''\n",
    "    t is a tree in newick format\n",
    "    return a tree in newick format\n",
    "    only include nodes that exist in dc_sample2species\n",
    "    '''\n",
    "    tree = Tree(t)\n",
    "    for l in tree.iter_leaves():\n",
    "        if l.name not in dc_sample2species:\n",
    "            l.delete()\n",
    "    return tree.write()\n",
    "    \n",
    "for line in open(filename).readlines():\n",
    "    n,t = line.split()\n",
    "    new_t = removeNodes(t,dc_sample2species)\n",
    "    fout.write(n+'\\t'+new_t+'\\n')\n",
    "fout.close()\n",
    "\n",
    "filename = '/home/xcao/w/2018Calephelis/20190101rawsoni/wholeGenomeGeneTrees/wholeGap7.RAxMLbestTree.all.26'\n",
    "txts = open(filename).readlines()\n",
    "pool = Pool(32)\n",
    "df_tree = pd.DataFrame(pool.map(processTXT,txts))\n",
    "pool.close()\n",
    "\n",
    "species_checking = ['Calephelis arizonensis','Calephelis freemani','Calephelis nuevoleon','Calephelis rawsoni']\n",
    "for species in species_checking:\n",
    "    print('trees with samples in one monotree for',species, sum(df_tree[species]))\n",
    "\n",
    "#get trees that separate all four species. No tree\n",
    "df_good = df_tree[df_tree['Calephelis arizonensis'] & df_tree['Calephelis freemani'] & df_tree['Calephelis nuevoleon'] & df_tree['Calephelis rawsoni']]\n",
    "\n",
    "outfolder = '/home/xcao/w/2018Calephelis/20190101rawsoni/wholeGenomeGeneTrees/wholeGap7.RAxML.26.keep/'\n",
    "for n,row in df_good.iterrows():\n",
    "    open(outfolder+row['tree_id'],'w').write(row['tree_txt'])\n",
    "    \n",
    "    \n",
    "#get trees that separate all four species. No tree\n",
    "df_good = df_tree[df_tree['Calephelis arizonensis'] & df_tree['Calephelis freemani'] & df_tree['Calephelis nuevoleon'] & df_tree['Calephelis rawsoni']]\n",
    "\n",
    "outfolder = '/home/xcao/w/2018Calephelis/20190101rawsoni/wholeGenomeGeneTrees/wholeGap7.RAxML.26.keep/'\n",
    "for n,row in df_good.iterrows():\n",
    "    tree=Tree(row['tree_txt'])\n",
    "    tree.unroot()\n",
    "    tree_txt = tree.write()\n",
    "    open(outfolder+row['tree_id'],'w').write(tree_txt)\n",
    "    \n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "plot\n",
    "```\n",
    "library(ggtree)\n",
    "library(ggrepel)\n",
    "library(ggplot2)\n",
    "\n",
    "tree <- read.tree('/home/xcao/w/2018Calephelis/20190101rawsoni/wholeGenomeGeneTrees/wholeGap7.RAxML.26.keep/52')\n",
    "\n",
    "xdf.sample2species = read.csv('/home/xcao/w/2018Calephelis/20190101rawsoni/wholeGenomeGeneTrees/id2species',sep = '\\t',stringsAsFactors = FALSE,header = FALSE,colClasses = 'character')\n",
    "colnames(xdf.sample2species) <- c('sample','species')\n",
    "\n",
    "p <- ggtree(tree,layout = 'daylight') \n",
    "p %<+% xdf.sample2species + geom_tiplab2(aes(angle=angle),hjust = 0.5,vjust=2) + geom_tippoint(aes(size=5,color=species,shape=species),alpha=0.7) + scale_shape_manual(values=c(15,16,17,15,16,17))+ theme(legend.position=\"right\")+ guides(colour = guide_legend(override.aes = list(size=5))) + ggtitle('test')+geom_nodelab(size = 5, col= \"red\")\n",
    "\n",
    "folder <- '/home/xcao/w/2018Calephelis/20190101rawsoni/wholeGenomeGeneTrees/wholeGap7.RAxML.26.keep/'\n",
    "files = c('52', '66', '2101')\n",
    "\n",
    "pdf('/home/xcao/w/2018Calephelis/20190101rawsoni/wholeGenomeGeneTrees/wholeGap7.RAxML.26.4SpeciesTogehter.pdf',width = 15, height = 9)\n",
    "for (t in files){\n",
    "  tree <- paste0(folder,t)\n",
    "  tree <- read.tree(tree)\n",
    "  p <- ggtree(tree,layout = 'daylight') \n",
    "  p <- p %<+% xdf.sample2species + geom_tiplab2(aes(angle=angle),hjust = 0.5,vjust=2) + geom_tippoint(aes(size=5,color=species,shape=species),alpha=0.7) + scale_shape_manual(values=c(15,16,17,15,16,17))+ theme(legend.position=\"right\")+ guides(colour = guide_legend(override.aes = list(size=5))) + ggtitle(t) + geom_nodelab(size = 5, col= \"red\")\n",
    "  print(p)\n",
    "}\n",
    "dev.off()\n",
    "\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "trees with samples in one monotree for Calephelis arizonensis 2893  \n",
    "trees with samples in one monotree for Calephelis freemani 51  \n",
    "trees with samples in one monotree for Calephelis nuevoleon 145  \n",
    "trees with samples in one monotree for Calephelis rawsoni 1324  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PCA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### run PCA for 266 samples, CDS, whole, Z and intron"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "don't forget to run the commands below\n",
    "```\n",
    "export PATH=$PATH:/home2/wli/local/EIG6.1.1/bin;\n",
    "module add gsl/1.15;\n",
    "module add openblas/intel/0.2.14;\n",
    "```\n",
    "\n",
    "generate running scripts\n",
    "```\n",
    "import os\n",
    "file_fastas = '''/work/biophysics/s185491/20181226Calephelis/Phylo/CDS/Calephelis266.CDS\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/intron/Calephelis266.intron.Gap100\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/whole/Calephelis266.wholeGap60\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/zChr/Calephelis266.zChr'''.split()\n",
    "folder_PCA = '/work/biophysics/s185491/20181226Calephelis/PCA/266/'\n",
    "\n",
    "txt = '''cd {folder_PCA}  &&  python /home2/s185491/p/xiaolongTools/WenlinTools.Python3/scripts/fasta2EIGinput.py {file_fasta}  &&  /home2/wli/local/EIG6.1.1/bin/smartpca.pl -i {filename}.geno -a {filename}.snp -b {filename}.ind -o {filename}.pca -p {filename}.plot -e {filename}.eval -l {filename}.log -snpweightoutname {filename}.snp_weights.txt &\n",
    "'''\n",
    "for file_fasta in file_fastas:\n",
    "    print(txt.format(folder_PCA=folder_PCA, file_fasta=file_fasta, filename=os.path.basename(file_fasta)))\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### plot PCA\n",
    "```\n",
    "\n",
    "filename <- 'C:\\\\Users\\\\ATPs\\\\OneDrive\\\\Lab\\\\UTSW\\\\2018Calephelis\\\\summary\\\\20181217Calephelis.xlsx'\n",
    "\n",
    "library(ggplot2)\n",
    "library(readxl)\n",
    "library(ggrepel)\n",
    "\n",
    "mydata <- read_excel(filename,sheet = 'Sheet1')\n",
    "taxon_table <- table(mydata$`Taxon name`)\n",
    "taxon_table[taxon_table>5]\n",
    "p <- c(0:25, 35,36,37,38,43,48:57,64:127)\n",
    "taxon_shape <- p[1:length(taxon_table)]\n",
    "names(taxon_shape) <- names(taxon_table)\n",
    "rename_taxon <- function(x,N=5){\n",
    "  #x is a taxon name. if count of x >= N, return x, else return 'other'\n",
    "  if (taxon_table[x]>=N) return(x)\n",
    "  else return(\"other\")\n",
    "}\n",
    "mydata$taxon <- sapply(mydata$`Taxon name`,rename_taxon)\n",
    "mydata$taxon.shape <- taxon_shape[mydata$`Taxon name`]\n",
    "\n",
    "ls_PCAs <- list(\n",
    "  c(\"PC1_266_zChr\", \"PC2_266_zChr\", \"PC3_266_zChr\"),\n",
    "  c(\"PC1_266_whole\",\"PC2_266_whole\", \"PC3_266_whole\"),\n",
    "  c(\"PC1_266_intron\", \"PC2_266_intron\", \"PC3_266_intron\"),\n",
    "  c(\"PC1_266_CDS\", \"PC2_266_CDS\", \"PC3_266_CDS\")\n",
    ")\n",
    "\n",
    "pdf('C:\\\\Users\\\\ATPs\\\\OneDrive\\\\Lab\\\\UTSW\\\\2018Calephelis\\\\PCA\\\\Calephelis266.zChr.whole.intron.CDS.pca.pdf',width = 36, height = 19)\n",
    "for (PCAs in ls_PCAs) {\n",
    "  p1 <- ggplot(mydata, aes_string(x=PCAs[1],y=PCAs[2])) + geom_point(aes(color=`Taxon name`, size = 5,shape=taxon.shape))  + geom_label_repel(aes(label=`DNA number`),label.size=NA,fill=NA) +scale_shape_identity() \n",
    "  p2 <- ggplot(mydata, aes_string(x=PCAs[1],y=PCAs[3])) + geom_point(aes(color=`Taxon name`, size = 5,shape=taxon.shape))  + geom_label_repel(aes(label=`DNA number`),label.size=NA,fill=NA)  +scale_shape_identity()\n",
    "  p3 <- ggplot(mydata, aes_string(x=PCAs[2],y=PCAs[3])) + geom_point(aes(color=`Taxon name`, size = 5,shape=taxon.shape))  + geom_label_repel(aes(label=`DNA number`),label.size=NA,fill=NA)  +scale_shape_identity()\n",
    "  print(p1)\n",
    "  print(p2)\n",
    "  print(p3)\n",
    "}\n",
    "\n",
    "dev.off()\n",
    "\n",
    "pdf('C:\\\\Users\\\\ATPs\\\\OneDrive\\\\Lab\\\\UTSW\\\\2018Calephelis\\\\PCA\\\\Calephelis266.zChr.whole.intron.CDS.pca.noLabel.pdf',width = 36, height = 19)\n",
    "for (PCAs in ls_PCAs) {\n",
    "  p1 <- ggplot(mydata, aes_string(x=PCAs[1],y=PCAs[2])) + geom_point(aes(color=`Taxon name`, size = 5,shape=taxon.shape))   +scale_shape_identity() \n",
    "  p2 <- ggplot(mydata, aes_string(x=PCAs[1],y=PCAs[3])) + geom_point(aes(color=`Taxon name`, size = 5,shape=taxon.shape))    +scale_shape_identity()\n",
    "  p3 <- ggplot(mydata, aes_string(x=PCAs[2],y=PCAs[3])) + geom_point(aes(color=`Taxon name`, size = 5,shape=taxon.shape))    +scale_shape_identity()\n",
    "  print(p1)\n",
    "  print(p2)\n",
    "  print(p3)\n",
    "}\n",
    "\n",
    "dev.off()\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ru PCA for 35 C. rawsoni group"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "run\n",
    "```\n",
    "import os\n",
    "file_fastas = '''/work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/Calephelis35.CDS\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/Calephelis35.intron\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/Calephelis35.mito\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/Calephelis35.wholeGap7\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/Calephelis35.zChr\n",
    "'''.split()\n",
    "folder_PCA = '/work/biophysics/s185491/20181226Calephelis/PCA/20190101rawsoni/'\n",
    "\n",
    "txt = '''cd {folder_PCA}  &&  python /home2/s185491/p/xiaolongTools/WenlinTools.Python3/scripts/fasta2EIGinput.py {file_fasta}  &&  /home2/wli/local/EIG6.1.1/bin/smartpca.pl -i {filename}.geno -a {filename}.snp -b {filename}.ind -o {filename}.pca -p {filename}.plot -e {filename}.eval -l {filename}.log -snpweightoutname {filename}.snp_weights.txt &\n",
    "'''\n",
    "for file_fasta in file_fastas:\n",
    "    print(txt.format(folder_PCA=folder_PCA, file_fasta=file_fasta, filename=os.path.basename(file_fasta)))\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "plot\n",
    "```\n",
    "filename <- 'C:\\\\Users\\\\ATPs\\\\OneDrive\\\\Lab\\\\UTSW\\\\2018Calephelis\\\\summary\\\\20181217Calephelis.xlsx'\n",
    "\n",
    "library(ggplot2)\n",
    "library(readxl)\n",
    "library(ggrepel)\n",
    "library(tidyr)\n",
    "\n",
    "mydata1 <- read_excel(filename,sheet = 'Sheet1')\n",
    "taxon_table <- table(mydata1$`Taxon name`)\n",
    "taxon_table[taxon_table>5]\n",
    "p <- c(0:25, 35,36,37,38,43,48:57,64:127)\n",
    "taxon_shape <- p[1:length(taxon_table)]\n",
    "names(taxon_shape) <- names(taxon_table)\n",
    "rename_taxon <- function(x,N=5){\n",
    "  #x is a taxon name. if count of x >= N, return x, else return 'other'\n",
    "  if (taxon_table[x]>=N) return(x)\n",
    "  else return(\"other\")\n",
    "}\n",
    "mydata1$taxon <- sapply(mydata1$`new_Taxon_name`,rename_taxon)\n",
    "mydata1$taxon.shape <- taxon_shape[mydata1$`new_Taxon_name`]\n",
    "\n",
    "PCA_head.str <- \"PC1_35_zChr\tPC2_35_zChr\tPC3_35_zChr\tPC1_35_whole\tPC2_35_whole\tPC3_35_whole\tPC1_35_intron\tPC2_35_intron\tPC3_35_intron\tPC1_35_CDS\tPC2_35_CDS\tPC3_35_CDS\tPC1_35_mito\tPC2_35_mito\tPC3_35_mito\"\n",
    "PCA_head.vector <- strsplit(PCA_head.str,split = '\\t')[[1]]\n",
    "\n",
    "\n",
    "pdf('C:\\\\Users\\\\ATPs\\\\OneDrive\\\\Lab\\\\UTSW\\\\2018Calephelis\\\\PCA\\\\Calephelis35.zChr.whole.intron.CDS.mito.pca.pdf',width = 18, height = 9)\n",
    "for (n in 1:(length(PCA_head.vector)/3)) {\n",
    "  PCAs <- PCA_head.vector[(n*3-2):(n*3)]\n",
    "  mydata <- drop_na(mydata1,PCAs[1])\n",
    "  p1 <- ggplot(mydata, aes_string(x=PCAs[1],y=PCAs[2])) + geom_point(aes(color=`new_Taxon_name`, size = 5,shape=new_Taxon_name))  + geom_label_repel(aes(label=`DNA number`),label.size=NA,fill=NA) + scale_shape_manual(values = rep(15:19, 6)) \n",
    "  p2 <- ggplot(mydata, aes_string(x=PCAs[1],y=PCAs[3])) + geom_point(aes(color=`new_Taxon_name`, size = 5,shape=new_Taxon_name))  + geom_label_repel(aes(label=`DNA number`),label.size=NA,fill=NA)  + scale_shape_manual(values = rep(15:19, 6))\n",
    "  p3 <- ggplot(mydata, aes_string(x=PCAs[2],y=PCAs[3])) + geom_point(aes(color=`new_Taxon_name`, size = 5,shape=new_Taxon_name))  + geom_label_repel(aes(label=`DNA number`),label.size=NA,fill=NA)  + scale_shape_manual(values = rep(15:19, 6))\n",
    "  figures = list(p1,p2,p3)\n",
    "  for (p in figures) print(p + guides(color = guide_legend(override.aes = list(size=5))))\n",
    "}\n",
    "\n",
    "dev.off()\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PCA 35 rawsoni with no outgroup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "file_fastas = '''/work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/Calephelis35.CDS\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/Calephelis35.intron\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/Calephelis35.mito\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/Calephelis35.wholeGap7\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/Calephelis35.zChr\n",
    "'''.split()\n",
    "txt = '''cd {folder_PCA}  &&  python /home2/s185491/p/xiaolongTools/WenlinTools.Python3/scripts/fasta2EIGinput.py {file_fasta}  &&  /home2/wli/local/EIG6.1.1/bin/smartpca.pl -i {filename}.geno -a {filename}.snp -b {filename}.ind -o {filename}.pca -p {filename}.plot -e {filename}.eval -l {filename}.log -snpweightoutname {filename}.snp_weights.txt &\n",
    "'''\n",
    "\n",
    "folder_PCA = '/work/biophysics/s185491/20181226Calephelis/PCA/20190106rawsoni_noOutgroup/'\n",
    "outgroups = ['5951', '18033G04', '15098B05', '16101C08', '6994', '5254', '7144', '16101B02', '3505']\n",
    "for file_fasta in file_fastas:\n",
    "    basename = os.path.basename(file_fasta)\n",
    "    outfile = os.path.join(folder_PCA, basename.replace('Calephelis35', 'Calephelis26'))\n",
    "    fout = open(outfile,'w')\n",
    "    for s in SeqIO.parse(file_fasta,'fasta'):\n",
    "        if s.id not in outgroups:\n",
    "            fout.write('>'+s.id+'\\n'+str(s.seq)+'\\n')\n",
    "    fout.close()\n",
    "    os.system(txt.format(folder_PCA=folder_PCA, file_fasta=outfile, filename=os.path.basename(outfile)))\n",
    "\n",
    "\n",
    "```\n",
    "\n",
    "\n",
    "```\n",
    "filename <- 'C:\\\\Users\\\\ATPs\\\\OneDrive\\\\Lab\\\\UTSW\\\\2018Calephelis\\\\summary\\\\20181217Calephelis.xlsx'\n",
    "\n",
    "library(ggplot2)\n",
    "library(readxl)\n",
    "library(ggrepel)\n",
    "library(tidyr)\n",
    "\n",
    "mydata1 <- read_excel(filename,sheet = 'Sheet1')\n",
    "taxon_table <- table(mydata1$`Taxon name`)\n",
    "taxon_table[taxon_table>5]\n",
    "p <- c(0:25, 35,36,37,38,43,48:57,64:127)\n",
    "taxon_shape <- p[1:length(taxon_table)]\n",
    "names(taxon_shape) <- names(taxon_table)\n",
    "rename_taxon <- function(x,N=5){\n",
    "  #x is a taxon name. if count of x >= N, return x, else return 'other'\n",
    "  if (taxon_table[x]>=N) return(x)\n",
    "  else return(\"other\")\n",
    "}\n",
    "mydata1$taxon <- sapply(mydata1$`new_Taxon_name`,rename_taxon)\n",
    "mydata1$taxon.shape <- taxon_shape[mydata1$`new_Taxon_name`]\n",
    "\n",
    "PCA_head.str <- \"PC1_26_zChr\tPC2_26_zChr\tPC3_26_zChr\tPC1_26_whole\tPC2_26_whole\tPC3_26_whole\tPC1_26_intron\tPC2_26_intron\tPC3_26_intron\tPC1_26_CDS\tPC2_26_CDS\tPC3_26_CDS\tPC1_26_mito\tPC2_26_mito\tPC3_26_mito\n",
    "\"\n",
    "PCA_head.vector <- strsplit(PCA_head.str,split = '\\t')[[1]]\n",
    "\n",
    "\n",
    "pdf('C:\\\\Users\\\\ATPs\\\\OneDrive\\\\Lab\\\\UTSW\\\\2018Calephelis\\\\PCA\\\\Calephelis26.zChr.whole.intron.CDS.mito.pca.pdf',width = 18, height = 9)\n",
    "for (n in 1:(length(PCA_head.vector)/3)) {\n",
    "  PCAs <- PCA_head.vector[(n*3-2):(n*3)]\n",
    "  mydata <- drop_na(mydata1,PCAs[1])\n",
    "  p1 <- ggplot(mydata, aes_string(x=PCAs[1],y=PCAs[2])) + geom_point(aes(color=`new_Taxon_name`, size = 5,shape=new_Taxon_name))  + geom_label_repel(aes(label=`DNA number`),label.size=NA,fill=NA) + scale_shape_manual(values = rep(15:19, 6)) \n",
    "  p2 <- ggplot(mydata, aes_string(x=PCAs[1],y=PCAs[3])) + geom_point(aes(color=`new_Taxon_name`, size = 5,shape=new_Taxon_name))  + geom_label_repel(aes(label=`DNA number`),label.size=NA,fill=NA)  + scale_shape_manual(values = rep(15:19, 6))\n",
    "  p3 <- ggplot(mydata, aes_string(x=PCAs[2],y=PCAs[3])) + geom_point(aes(color=`new_Taxon_name`, size = 5,shape=new_Taxon_name))  + geom_label_repel(aes(label=`DNA number`),label.size=NA,fill=NA)  + scale_shape_manual(values = rep(15:19, 6))\n",
    "  figures = list(p1,p2,p3)\n",
    "  for (p in figures) print(p + guides(color = guide_legend(override.aes = list(size=5))))\n",
    "}\n",
    "\n",
    "dev.off()\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Structure"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 266 sequences test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### convert sequences to structure input\n",
    "\n",
    "```\n",
    "files = '''\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/CDS/Calephelis266.CDS.split5.0\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/CDS/Calephelis266.CDS.split5.1\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/CDS/Calephelis266.CDS.split5.2\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/CDS/Calephelis266.CDS.split5.3\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/CDS/Calephelis266.CDS.split5.4\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/intron/Calephelis266.intron.Gap100.split5.0\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/intron/Calephelis266.intron.Gap100.split5.1\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/intron/Calephelis266.intron.Gap100.split5.2\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/intron/Calephelis266.intron.Gap100.split5.3\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/intron/Calephelis266.intron.Gap100.split5.4\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/mito/Calephelis266.mito\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/whole/Calephelis266.wholeGap60.split5.0\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/whole/Calephelis266.wholeGap60.split5.1\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/whole/Calephelis266.wholeGap60.split5.2\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/whole/Calephelis266.wholeGap60.split5.3\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/whole/Calephelis266.wholeGap60.split5.4\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/zChr/Calephelis266.zChr.split5.0\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/zChr/Calephelis266.zChr.split5.1\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/zChr/Calephelis266.zChr.split5.2\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/zChr/Calephelis266.zChr.split5.3\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/zChr/Calephelis266.zChr.split5.4\n",
    "'''.split()\n",
    "\n",
    "\n",
    "commandline = 'python /home2/s185491/p/xiaolongTools/utils/STRUCTURE/fasta2STRUCTUREinput_haploid_nt.py -i {file_in} -o {file_out} -t 56 -g {gap} -N 4 -H True'\n",
    "outfolder = '/work/biophysics/s185491/20181226Calephelis/structure/266/'\n",
    "import os\n",
    "\n",
    "for file in files:\n",
    "    if 'CDS.split' in file:\n",
    "        gap = 0.8\n",
    "    elif 'intron.Gap' in file:\n",
    "        gap = 0.85\n",
    "    elif 'wholeGap60.split' in file:\n",
    "        gap = 0.9\n",
    "    elif 'zChr.split' in file:\n",
    "        gap = 0.8\n",
    "    elif 'Calephelis266.mito' in file:\n",
    "        gap = 0.5\n",
    "    print(commandline.format(file_in = file,gap=gap, file_out = os.path.join(outfolder, os.path.basename(file))))\n",
    "    \n",
    "    \n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### run Structure"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "files = '''\n",
    "/work/biophysics/s185491/20181226Calephelis/structure/266/Calephelis266.CDS.split5.0\n",
    "/work/biophysics/s185491/20181226Calephelis/structure/266/Calephelis266.CDS.split5.1\n",
    "/work/biophysics/s185491/20181226Calephelis/structure/266/Calephelis266.CDS.split5.2\n",
    "/work/biophysics/s185491/20181226Calephelis/structure/266/Calephelis266.CDS.split5.3\n",
    "/work/biophysics/s185491/20181226Calephelis/structure/266/Calephelis266.CDS.split5.4\n",
    "/work/biophysics/s185491/20181226Calephelis/structure/266/Calephelis266.intron.Gap100.split5.0\n",
    "/work/biophysics/s185491/20181226Calephelis/structure/266/Calephelis266.intron.Gap100.split5.1\n",
    "/work/biophysics/s185491/20181226Calephelis/structure/266/Calephelis266.intron.Gap100.split5.2\n",
    "/work/biophysics/s185491/20181226Calephelis/structure/266/Calephelis266.intron.Gap100.split5.3\n",
    "/work/biophysics/s185491/20181226Calephelis/structure/266/Calephelis266.intron.Gap100.split5.4\n",
    "/work/biophysics/s185491/20181226Calephelis/structure/266/Calephelis266.mito\n",
    "/work/biophysics/s185491/20181226Calephelis/structure/266/Calephelis266.wholeGap60.split5.0\n",
    "/work/biophysics/s185491/20181226Calephelis/structure/266/Calephelis266.wholeGap60.split5.1\n",
    "/work/biophysics/s185491/20181226Calephelis/structure/266/Calephelis266.wholeGap60.split5.2\n",
    "/work/biophysics/s185491/20181226Calephelis/structure/266/Calephelis266.wholeGap60.split5.3\n",
    "/work/biophysics/s185491/20181226Calephelis/structure/266/Calephelis266.wholeGap60.split5.4\n",
    "/work/biophysics/s185491/20181226Calephelis/structure/266/Calephelis266.zChr.split5.0\n",
    "/work/biophysics/s185491/20181226Calephelis/structure/266/Calephelis266.zChr.split5.1\n",
    "/work/biophysics/s185491/20181226Calephelis/structure/266/Calephelis266.zChr.split5.2\n",
    "/work/biophysics/s185491/20181226Calephelis/structure/266/Calephelis266.zChr.split5.3\n",
    "/work/biophysics/s185491/20181226Calephelis/structure/266/Calephelis266.zChr.split5.4\n",
    "'''.split()\n",
    "\n",
    "snps = '''\n",
    "32468\n",
    "32057\n",
    "32230\n",
    "32084\n",
    "31706\n",
    "62363\n",
    "62500\n",
    "62052\n",
    "62581\n",
    "62456\n",
    "3731\n",
    "74496\n",
    "74293\n",
    "74093\n",
    "73866\n",
    "74081\n",
    "23622\n",
    "23627\n",
    "23443\n",
    "23684\n",
    "23509\n",
    "'''.split()\n",
    "\n",
    "commandline = 'python /home2/s185491/p/xiaolongTools/utils/STRUCTURE/runSTRUCTUREwithSTRUCTUREinput.py -i {file_in} -I 266 -L {snp_len} -s 1000 -P 2,3,4,5,6,7,8,9,10,11,12,13,14,15 -o /work/biophysics/s185491/20181226Calephelis/structure/266/cmds/jobs'\n",
    "outfolder = '/work/biophysics/s185491/20181226Calephelis/structure/266/'\n",
    "\n",
    "for file_in, snp_len in zip(files,snps):\n",
    "    print(commandline.format(file_in=file_in, snp_len=snp_len))\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "python /home2/s185491/p/xiaolongTools/utils/splitFiles2Nparts.py -i /work/biophysics/s185491/20181226Calephelis/structure/266/cmds/jobs -N 6"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### collect data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "convert data in two format, one is good for R or python, one is good for CLUMPP\n",
    "```\n",
    "import os\n",
    "import glob\n",
    "import sys\n",
    "sys.path.append('/home2/s185491/p/xiaolongTools/utils/STRUCTURE')\n",
    "import extractStructureTableFromCommandlineOutput\n",
    "folder_structure = '/work/biophysics/s185491/20181226Calephelis/structure/266/'\n",
    "outfolder = '/work/biophysics/s185491/20181226Calephelis/structure/266/results/'\n",
    "files_structure = glob.glob(folder_structure +'runSTRUCTURE*/*_f')\n",
    "for f in files_structure:\n",
    "    key = os.path.basename(f)\n",
    "    outfile = os.path.join(outfolder, key)\n",
    "\n",
    "outfolder= '/work/biophysics/s185491/20181226Calephelis/structure/266/results2/'\n",
    "dc_results = {}\n",
    "for f in files_structure:\n",
    "    key = os.path.basename(f)\n",
    "    outfile = os.path.join(outfolder, key)\n",
    "    dc_results[key] = extractStructureTableFromCommandlineOutput.extractStructureTableFromCommandlineOutput(filename=f, output=outfile, processLineForR=0)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### plot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "library(readxl)\n",
    "library(tidyr)\n",
    "library(dplyr)\n",
    "library(ggplot2)\n",
    "library(gridExtra)\n",
    "\n",
    "folder_files <- '/home/xcao/w/2018Calephelis/20181231Structure/results'\n",
    "result_files <- list.files(folder_files,full.names = FALSE)\n",
    "basetypes <- c(\"intron.Gap100\",\"CDS\",\"mito\",\"wholeGap60\",\"zChr\")\n",
    "\n",
    "f.getbasetype <- function(x){\n",
    "  for (keyword in basetypes){\n",
    "    if (grepl(keyword,x)) return(keyword)\n",
    "  }\n",
    "}\n",
    "\n",
    "f.getSplitgroup <- function(x){\n",
    "  if (grepl('.split',x)) {\n",
    "    return(as.integer(gsub(\"(.*split\\\\d\\\\.)|(\\\\.STRUCTUREresult.*)\",\"\",x)))\n",
    "  } else return(0)\n",
    "}\n",
    "\n",
    "f.getRandint <- function(x){\n",
    "  #x=\"Calephelis266.zChr.split5.4.STRUCTUREresult.s1000.pop8_f\" return 1000\n",
    "  return(as.integer(gsub('(.*STRUCTUREresult.s)|(.pop.*_f)','',x)))\n",
    "}\n",
    "\n",
    "f.getPopulation <- function(x) {\n",
    "  return(as.integer(gsub('(.*STRUCTUREresult.s\\\\d*.pop)|(_f)','',x)))\n",
    "}\n",
    "\n",
    "df_files <- data.frame(baseType=sapply(result_files, f.getbasetype),\n",
    "                       splitID=sapply(result_files, f.getSplitgroup),\n",
    "                       seed=sapply(result_files, f.getRandint),\n",
    "                       population=sapply(result_files,f.getPopulation),\n",
    "                       filename=file.path(folder_files,result_files), stringsAsFactors = FALSE)\n",
    "\n",
    "df_files <- df_files[order(df_files$baseType, df_files$population,df_files$seed,df_files$splitID),]\n",
    "\n",
    "F.renameHead <- function(x){\n",
    "  x[1] = 'sample'\n",
    "  x[2:length(x)] = paste0('pop',1:(length(x)-1))\n",
    "  return(x)\n",
    "}\n",
    "\n",
    "filename2 <- '/home/xcao/w/2018Calephelis/summary/20181217Calephelis.xlsx'\n",
    "xdf_info <- read_excel(filename2,sheet = 'Sheet1')\n",
    "xdf_info <- xdf_info[!is.na(xdf_info$order_266mitoTree),]\n",
    "# v_id2Treename <- xdf_info$Treename\n",
    "# names(v_id2Treename) <- sapply(xdf_info$ID, function(x)substring(x,2))\n",
    "v_id2species <- xdf_info$new_Taxon_name\n",
    "names(v_id2species) <- sapply(xdf_info$ID, function(x)substring(x,2))\n",
    "species_in_order <- xdf_info[order(xdf_info$order_266mitoTree),]$ID\n",
    "species_in_order <- substr(species_in_order,2,length(species_in_order))\n",
    "\n",
    "\n",
    "oneline = df_files[1,]\n",
    "xdf_s <- read.csv(oneline$filename,header = FALSE,sep = '\\t', stringsAsFactors = FALSE)\n",
    "colnames(xdf_s) <- F.renameHead(colnames(xdf_s))\n",
    "xdf_s$species <- v_id2species[xdf_s$sample]\n",
    "#xdf_s$treename <- v_id2Treename[xdf_s$sample]\n",
    "xdf_s.clear <- gather(xdf_s,key = 'pop',value = 'fraction',-sample,-species)\n",
    "xdf_s.clear <- xdf_s.clear[order(xdf_s.clear$species,xdf_s.clear$pop,-xdf_s.clear$fraction),]\n",
    "\n",
    "p <- ggplot(xdf_s.clear,aes(x=sample,y=fraction,fill=pop)) + geom_bar(stat = 'identity') + scale_x_discrete(limits=species_in_order)\n",
    "\n",
    "\n",
    "\n",
    "F.plotOneline <- function(oneline,xtext=FALSE){\n",
    "  xdf_s <- read.csv(oneline$filename,header = FALSE,sep = '\\t', stringsAsFactors = FALSE)\n",
    "  colnames(xdf_s) <- F.renameHead(colnames(xdf_s))\n",
    "  xdf_s$species <- v_id2species[xdf_s$sample]\n",
    "  #xdf_s$treename <- v_id2Treename[xdf_s$sample]\n",
    "  xdf_s.clear <- gather(xdf_s,key = 'pop',value = 'fraction',-sample,-species)\n",
    "  xdf_s.clear <- xdf_s.clear[order(xdf_s.clear$species,xdf_s.clear$pop,-xdf_s.clear$fraction),]\n",
    "  \n",
    "  p <- ggplot(xdf_s.clear,aes(x=sample,y=fraction,fill=pop)) + geom_bar(stat = 'identity') + scale_x_discrete(limits=species_in_order) + ylab(paste0('seed',oneline$seed,'_',oneline$baseType,'_K=',oneline$population,\" \",oneline$splitID))\n",
    "  if (xtext) {\n",
    "    p <- p + theme(legend.position=\"none\", axis.text.x= element_text(angle = 90, hjust = 1), axis.title.x = element_blank())\n",
    "  }\n",
    "  else {\n",
    "    p <-  p + theme(legend.position=\"none\", axis.text.x= element_blank(), axis.title.x = element_blank())\n",
    "    }\n",
    " return(p)\n",
    "}\n",
    "\n",
    "F.plotMultipleLines <- function(filelines,xtext=FALSE){\n",
    "  p <- list()\n",
    "  for (i in 1:nrow(filelines)){\n",
    "    p[[i]] <- F.plotOneline(filelines[i,],xtext = xtext)\n",
    "  }\n",
    "  #p[[i+1]] <- F.plotOneline(filelines[i+1,], xtext = TRUE)\n",
    "  return(grid.arrange(grobs=p,ncol=1))\n",
    "}\n",
    "\n",
    "filelines = df_files[df_files$baseType == 'mito',]\n",
    "F.plotMultipleLines(filelines = filelines)\n",
    "\n",
    "filelines = df_files[df_files$baseType == 'mito',]\n",
    "F.plotMultipleLines(filelines = filelines,xtext = TRUE)\n",
    "\n",
    "filelines = df_files[df_files$baseType == 'zChr',]\n",
    "F.plotMultipleLines(filelines = filelines)\n",
    "\n",
    "filelines = df_files[df_files$baseType == 'CDS',]\n",
    "F.plotMultipleLines(filelines = filelines)\n",
    "\n",
    "filelines = df_files[df_files$baseType == 'CDS' &df_files$splitID==0,]\n",
    "F.plotMultipleLines(filelines = filelines,xtext = TRUE)\n",
    "\n",
    "\n",
    "filelines = df_files[df_files$baseType == 'wholeGap60',]\n",
    "F.plotMultipleLines(filelines = filelines)\n",
    "\n",
    "filelines = df_files[df_files$baseType == 'intron.Gap100',]\n",
    "F.plotMultipleLines(filelines = filelines)\n",
    "\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 35 rawsoni"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### convert sequence to structure input\n",
    "```\n",
    "files = '''\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/Calephelis35.CDS\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/Calephelis35.intron\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/Calephelis35.mito\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/Calephelis35.wholeGap7\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/Calephelis35.zChr\n",
    "'''.split()\n",
    "\n",
    "\n",
    "commandline = 'python /home2/s185491/p/xiaolongTools/utils/STRUCTURE/fasta2STRUCTUREinput_haploid_nt.py -i {file_in} -o {file_out} -t 56 -g {gap} -N 4 -H True'\n",
    "outfolder = '/work/biophysics/s185491/20181226Calephelis/structure/35rawsoni/'\n",
    "import os\n",
    "\n",
    "for file in files:\n",
    "    if 'CDS' in file:\n",
    "        gap = 0.8\n",
    "    elif 'intron' in file:\n",
    "        gap = 0.85\n",
    "    elif 'whole' in file:\n",
    "        gap = 0.9\n",
    "    elif 'zChr' in file:\n",
    "        gap = 0.8\n",
    "    elif 'mito' in file:\n",
    "        gap = 0.5\n",
    "    print(commandline.format(file_in = file,gap=gap, file_out = os.path.join(outfolder, os.path.basename(file))))\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### run structure\n",
    "```\n",
    "files = '''/work/biophysics/s185491/20181226Calephelis/structure/35rawsoni/Calephelis35.CDS\n",
    "/work/biophysics/s185491/20181226Calephelis/structure/35rawsoni/Calephelis35.intron\n",
    "/work/biophysics/s185491/20181226Calephelis/structure/35rawsoni/Calephelis35.mito\n",
    "/work/biophysics/s185491/20181226Calephelis/structure/35rawsoni/Calephelis35.wholeGap7\n",
    "/work/biophysics/s185491/20181226Calephelis/structure/35rawsoni/Calephelis35.zChr\n",
    "'''.split()\n",
    "\n",
    "snps = '''\n",
    "168647\n",
    "529582\n",
    "2196\n",
    "780165\n",
    "109709\n",
    "'''.split()\n",
    "\n",
    "commandline = 'python /home2/s185491/p/xiaolongTools/utils/STRUCTURE/runSTRUCTUREwithSTRUCTUREinput.py -i {file_in} -I 35 -L {snp_len} -s 1000,2000,3000 -P 2,3,4,5,6 -o /work/biophysics/s185491/20181226Calephelis/structure/35rawsoni/jobs'\n",
    "outfolder = '/work/biophysics/s185491/20181226Calephelis/structure/35rawsoni/'\n",
    "import os\n",
    "for file_in, snp_len in zip(files,snps):\n",
    "    os.system(commandline.format(file_in=file_in, snp_len=snp_len))\n",
    "```\n",
    "modify the jobs file and submit"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### collect data\n",
    "```\n",
    "import os\n",
    "import glob\n",
    "import sys\n",
    "sys.path.append('/home2/s185491/p/xiaolongTools/utils/STRUCTURE')\n",
    "import extractStructureTableFromCommandlineOutput\n",
    "folder_structure = '/work/biophysics/s185491/20181226Calephelis/structure/35rawsoni/'\n",
    "outfolder = '/work/biophysics/s185491/20181226Calephelis/structure/35rawsoni/result/'\n",
    "files_structure = glob.glob(folder_structure +'runSTRUCTURE*/*_f')\n",
    "for f in files_structure:\n",
    "    key = os.path.basename(f)\n",
    "    outfile = os.path.join(outfolder, key)\n",
    "    extractStructureTableFromCommandlineOutput.extractStructureTableFromCommandlineOutput(filename=f, output=outfile, processLineForR=1)\n",
    "\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### plot\n",
    "```\n",
    "library(readxl)\n",
    "library(tidyr)\n",
    "library(dplyr)\n",
    "library(ggplot2)\n",
    "library(gridExtra)\n",
    "\n",
    "folder_files <- '/home/xcao/w/2018Calephelis/20181231Structure/35/result/'\n",
    "result_files <- list.files(folder_files,full.names = FALSE)\n",
    "basetypes <- c(\"intron\",\"CDS\",\"mito\",\"wholeGap7\",\"zChr\")\n",
    "\n",
    "f.getbasetype <- function(x){\n",
    "  for (keyword in basetypes){\n",
    "    if (grepl(keyword,x)) return(keyword)\n",
    "  }\n",
    "}\n",
    "\n",
    "f.getSplitgroup <- function(x){\n",
    "  if (grepl('.split',x)) {\n",
    "    return(as.integer(gsub(\"(.*split\\\\d\\\\.)|(\\\\.STRUCTUREresult.*)\",\"\",x)))\n",
    "  } else return(0)\n",
    "}\n",
    "\n",
    "f.getRandint <- function(x){\n",
    "  #x=\"Calephelis266.zChr.split5.4.STRUCTUREresult.s1000.pop8_f\" return 1000\n",
    "  return(as.integer(gsub('(.*STRUCTUREresult.s)|(.pop.*_f)','',x)))\n",
    "}\n",
    "\n",
    "f.getPopulation <- function(x) {\n",
    "  return(as.integer(gsub('(.*STRUCTUREresult.s\\\\d*.pop)|(_f)','',x)))\n",
    "}\n",
    "\n",
    "df_files <- data.frame(baseType=sapply(result_files, f.getbasetype),\n",
    "                       splitID=sapply(result_files, f.getSplitgroup),\n",
    "                       seed=sapply(result_files, f.getRandint),\n",
    "                       population=sapply(result_files,f.getPopulation),\n",
    "                       filename=file.path(folder_files,result_files), stringsAsFactors = FALSE)\n",
    "\n",
    "df_files <- df_files[order(df_files$baseType, df_files$population,df_files$seed,df_files$splitID),]\n",
    "\n",
    "F.renameHead <- function(x){\n",
    "  x[1] = 'sample'\n",
    "  x[2:length(x)] = paste0('pop',1:(length(x)-1))\n",
    "  return(x)\n",
    "}\n",
    "\n",
    "filename2 <- '/home/xcao/w/2018Calephelis/summary/20181217Calephelis.xlsx'\n",
    "xdf_info <- read_excel(filename2,sheet = 'Sheet1')\n",
    "xdf_info <- xdf_info[!is.na(xdf_info$Filter_rawsoni),]\n",
    "# v_id2Treename <- xdf_info$Treename\n",
    "# names(v_id2Treename) <- sapply(xdf_info$ID, function(x)substring(x,2))\n",
    "v_id2species <- xdf_info$new_Taxon_name\n",
    "names(v_id2species) <- sapply(xdf_info$ID, function(x)substring(x,2))\n",
    "species_in_order <- xdf_info[order(xdf_info$order_266mitoTree),]$ID\n",
    "species_in_order <- substr(species_in_order,2,length(species_in_order))\n",
    "\n",
    "\n",
    "oneline = df_files[1,]\n",
    "xdf_s <- read.csv(oneline$filename,header = FALSE,sep = '\\t', stringsAsFactors = FALSE)\n",
    "colnames(xdf_s) <- F.renameHead(colnames(xdf_s))\n",
    "xdf_s$species <- v_id2species[xdf_s$sample]\n",
    "#xdf_s$treename <- v_id2Treename[xdf_s$sample]\n",
    "xdf_s.clear <- gather(xdf_s,key = 'pop',value = 'fraction',-sample,-species)\n",
    "xdf_s.clear <- xdf_s.clear[order(xdf_s.clear$species,xdf_s.clear$pop,-xdf_s.clear$fraction),]\n",
    "\n",
    "p <- ggplot(xdf_s.clear,aes(x=sample,y=fraction,fill=pop)) + geom_bar(stat = 'identity') + scale_x_discrete(limits=species_in_order)\n",
    "\n",
    "\n",
    "\n",
    "F.plotOneline <- function(oneline,xtext=FALSE){\n",
    "  xdf_s <- read.csv(oneline$filename,header = FALSE,sep = '\\t', stringsAsFactors = FALSE)\n",
    "  colnames(xdf_s) <- F.renameHead(colnames(xdf_s))\n",
    "  xdf_s$species <- v_id2species[xdf_s$sample]\n",
    "  #xdf_s$treename <- v_id2Treename[xdf_s$sample]\n",
    "  xdf_s.clear <- gather(xdf_s,key = 'pop',value = 'fraction',-sample,-species)\n",
    "  xdf_s.clear <- xdf_s.clear[order(xdf_s.clear$species,xdf_s.clear$pop,-xdf_s.clear$fraction),]\n",
    "  \n",
    "  p <- ggplot(xdf_s.clear,aes(x=sample,y=fraction,fill=pop)) + geom_bar(stat = 'identity') + scale_x_discrete(limits=species_in_order) + ylab(paste0('seed',oneline$seed,'_',oneline$baseType,'_K=',oneline$population,\" \",oneline$splitID))\n",
    "  if (xtext) {\n",
    "    p <- p + theme(legend.position=\"none\", axis.text.x= element_text(angle = 90, hjust = 1), axis.title.x = element_blank())\n",
    "  }\n",
    "  else {\n",
    "    p <-  p + theme(legend.position=\"none\", axis.text.x= element_blank(), axis.title.x = element_blank())\n",
    "  }\n",
    "  return(p)\n",
    "}\n",
    "\n",
    "F.plotMultipleLines <- function(filelines,xtext=FALSE){\n",
    "  p <- list()\n",
    "  for (i in 1:nrow(filelines)){\n",
    "    p[[i]] <- F.plotOneline(filelines[i,],xtext = xtext)\n",
    "  }\n",
    "  #p[[i+1]] <- F.plotOneline(filelines[i+1,], xtext = TRUE)\n",
    "  return(grid.arrange(grobs=p,ncol=1))\n",
    "}\n",
    "\n",
    "#\"intron\",\"CDS\",\"mito\",\"wholeGap7\",\"zChr\"\n",
    "filelines = df_files[df_files$baseType == 'mito'& df_files$seed == 1000,]\n",
    "F.plotMultipleLines(filelines = filelines,xtext = TRUE)\n",
    "\n",
    "pdf('/home/xcao/w/2018Calephelis/20181231Structure/35/35rawsoni.pdf',width = 8, height = 12)\n",
    "filelines = df_files[df_files$baseType == 'mito' & df_files$seed == 1000,]\n",
    "F.plotMultipleLines(filelines = filelines)\n",
    "\n",
    "\n",
    "\n",
    "filelines = df_files[df_files$baseType == 'zChr'& df_files$seed == 1000,]\n",
    "F.plotMultipleLines(filelines = filelines)\n",
    "\n",
    "filelines = df_files[df_files$baseType == 'CDS'& df_files$seed == 1000,]\n",
    "F.plotMultipleLines(filelines = filelines)\n",
    "\n",
    "\n",
    "filelines = df_files[df_files$baseType == 'wholeGap7'& df_files$seed == 1000,]\n",
    "F.plotMultipleLines(filelines = filelines)\n",
    "\n",
    "filelines = df_files[df_files$baseType == 'intron'& df_files$seed == 1000,]\n",
    "F.plotMultipleLines(filelines = filelines)\n",
    "dev.off()\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### rawsoni no outgroup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "files = '''/work/biophysics/s185491/20181226Calephelis/structure/35rawsoni/Calephelis35.CDS\n",
    "/work/biophysics/s185491/20181226Calephelis/structure/35rawsoni/Calephelis35.intron\n",
    "/work/biophysics/s185491/20181226Calephelis/structure/35rawsoni/Calephelis35.mito\n",
    "/work/biophysics/s185491/20181226Calephelis/structure/35rawsoni/Calephelis35.wholeGap7\n",
    "/work/biophysics/s185491/20181226Calephelis/structure/35rawsoni/Calephelis35.zChr\n",
    "'''.split()\n",
    "\n",
    "snps = '''\n",
    "168647\n",
    "529582\n",
    "2196\n",
    "780165\n",
    "109709\n",
    "'''.split()\n",
    "\n",
    "filter1 = ['5951', '18033G04', '15098B05', '16101C08', '6994', '5254', '7144', '16101B02', '3505']\n",
    "filter2 = ['7144', '16101B02', '3505']\n",
    "outfolder1 = '/work/biophysics/s185491/20181226Calephelis/structure/rawsoni26/'\n",
    "outfolder2 = '/work/biophysics/s185491/20181226Calephelis/structure/rawsoni32/'\n",
    "\n",
    "commandline = 'python /home2/s185491/p/xiaolongTools/utils/STRUCTURE/runSTRUCTUREwithSTRUCTUREinput.py -i {file_in} -I {inputN} -L {snp_len} -s 1000,2000,3000 -P 2,3,4,5,6,7,8,9 -o /work/biophysics/s185491/20181226Calephelis/structure/rawsoni_jobs'\n",
    "\n",
    "import os\n",
    "for file_in, snp_len in zip(files,snps):\n",
    "    l = open(file_in).readlines()\n",
    "    file_out1 = os.path.join(outfolder1,os.path.basename(file_in).replace('Calephelis35','Calephelis26'))\n",
    "    file_out2 = os.path.join(outfolder2,os.path.basename(file_in).replace('Calephelis35','Calephelis32'))\n",
    "    fout1 = open(file_out1, 'w')\n",
    "    fout2 = open(file_out2, 'w')\n",
    "    for line in l:\n",
    "        if line.split('\\t')[0] not in filter1:\n",
    "            fout1.write(line)\n",
    "        if line.split('\\t')[0] not in filter2:\n",
    "            fout2.write(line)\n",
    "    fout1.close()\n",
    "    fout2.close()\n",
    "    os.system(commandline.format(file_in=file_out1, snp_len=snp_len, inputN=26))\n",
    "    os.system(commandline.format(file_in=file_out2, snp_len=snp_len, inputN=32))\n",
    "\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### collect data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "import os\n",
    "import glob\n",
    "import sys\n",
    "sys.path.append('/home2/s185491/p/xiaolongTools/utils/STRUCTURE')\n",
    "import extractStructureTableFromCommandlineOutput\n",
    "folder_structure = '/work/biophysics/s185491/20181226Calephelis/structure/rawsoni26/'\n",
    "outfolder = '/work/biophysics/s185491/20181226Calephelis/structure/rawsoni26/result/'\n",
    "files_structure = glob.glob(folder_structure +'runSTRUCTURE*/*_f')\n",
    "for f in files_structure:\n",
    "    key = os.path.basename(f)\n",
    "    outfile = os.path.join(outfolder, key)\n",
    "    extractStructureTableFromCommandlineOutput.extractStructureTableFromCommandlineOutput(filename=f, output=outfile, processLineForR=1)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### plot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### rawsoni 26 zChrCDS\n",
    "```\n",
    "files = '''\n",
    "/work/biophysics/s185491/20181226Calephelis/Phylo/20190101rawsoni/Calephelis35.zChrCDS\n",
    "'''.split()\n",
    "\n",
    "\n",
    "commandline = 'python /home2/s185491/p/xiaolongTools/utils/STRUCTURE/fasta2STRUCTUREinput_haploid_nt.py -i {file_in} -o {file_out} -t 56 -g {gap} -N 4 -H True'\n",
    "outfolder = '/work/biophysics/s185491/20181226Calephelis/structure/35rawsoni/'\n",
    "import os\n",
    "\n",
    "for file in files:\n",
    "    if 'CDS' in file:\n",
    "        gap = 0.8\n",
    "    elif 'intron' in file:\n",
    "        gap = 0.85\n",
    "    elif 'whole' in file:\n",
    "        gap = 0.9\n",
    "    elif 'zChr' in file:\n",
    "        gap = 0.8\n",
    "    elif 'mito' in file:\n",
    "        gap = 0.5\n",
    "    print(commandline.format(file_in = file,gap=gap, file_out = os.path.join(outfolder, os.path.basename(file))))\n",
    "\n",
    "\n",
    "files = '''/work/biophysics/s185491/20181226Calephelis/structure/35rawsoni/Calephelis35.zChrCDS\n",
    "'''.split()\n",
    "\n",
    "snps = '''\n",
    "3989\n",
    "'''.split()\n",
    "\n",
    "filter1 = ['5951', '18033G04', '15098B05', '16101C08', '6994', '5254', '7144', '16101B02', '3505']\n",
    "outfolder1 = '/work/biophysics/s185491/20181226Calephelis/structure/rawsoni26/'\n",
    "\n",
    "commandline = 'python /home2/s185491/p/xiaolongTools/utils/STRUCTURE/runSTRUCTUREwithSTRUCTUREinput.py -i {file_in} -I {inputN} -L {snp_len} -s 1000,2000,3000 -P 2,3,4,5,6,7,8,9 -o /work/biophysics/s185491/20181226Calephelis/structure/rawsoni_jobs2'\n",
    "\n",
    "import os\n",
    "for file_in, snp_len in zip(files,snps):\n",
    "    l = open(file_in).readlines()\n",
    "    file_out1 = os.path.join(outfolder1,os.path.basename(file_in).replace('Calephelis35','Calephelis26'))\n",
    "    fout1 = open(file_out1, 'w')\n",
    "    for line in l:\n",
    "        if line.split('\\t')[0] not in filter1:\n",
    "            fout1.write(line)\n",
    "\n",
    "    fout1.close()\n",
    "    os.system(commandline.format(file_in=file_out1, snp_len=snp_len, inputN=26))\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### plot 26 rawsoni\n",
    "```\n",
    "library(readxl)\n",
    "library(tidyr)\n",
    "library(dplyr)\n",
    "library(ggplot2)\n",
    "library(gridExtra)\n",
    "\n",
    "folder_files <- '/home/xcao/w/2018Calephelis/20181231Structure/26/result/'\n",
    "result_files <- list.files(folder_files,full.names = FALSE)\n",
    "basetypes <- c(\"intron\",\"CDS\",\"mito\",\"wholeGap7\",\"zChr\",\"zChrCDS\")\n",
    "\n",
    "f.getbasetype <- function(x){\n",
    "  return (strsplit(x,split = '\\\\.')[[1]][2])\n",
    "}\n",
    "\n",
    "f.getSplitgroup <- function(x){\n",
    "  if (grepl('.split',x)) {\n",
    "    return(as.integer(gsub(\"(.*split\\\\d*\\\\.)|(\\\\.STRUCTUREresult.*)\",\"\",x)))\n",
    "  } else return(0)\n",
    "}\n",
    "\n",
    "f.getRandint <- function(x){\n",
    "  #x=\"Calephelis266.zChr.split5.4.STRUCTUREresult.s1000.pop8_f\" return 1000\n",
    "  return(as.integer(gsub('(.*STRUCTUREresult.s)|(.pop.*_f)','',x)))\n",
    "}\n",
    "\n",
    "f.getPopulation <- function(x) {\n",
    "  return(as.integer(gsub('(.*STRUCTUREresult.s\\\\d*.pop)|(_f)','',x)))\n",
    "}\n",
    "\n",
    "df_files <- data.frame(baseType=sapply(result_files, f.getbasetype),\n",
    "                       splitID=sapply(result_files, f.getSplitgroup),\n",
    "                       seed=sapply(result_files, f.getRandint),\n",
    "                       population=sapply(result_files,f.getPopulation),\n",
    "                       filename=file.path(folder_files,result_files), stringsAsFactors = FALSE)\n",
    "\n",
    "df_files <- df_files[order(df_files$baseType, df_files$population,df_files$seed,df_files$splitID),]\n",
    "\n",
    "F.renameHead <- function(x){\n",
    "  x[1] = 'sample'\n",
    "  x[2:length(x)] = paste0('pop',1:(length(x)-1))\n",
    "  return(x)\n",
    "}\n",
    "\n",
    "filename2 <- '/home/xcao/w/2018Calephelis/summary/20181217Calephelis.xlsx'\n",
    "xdf_info <- read_excel(filename2,sheet = 'Sheet1')\n",
    "xdf_info <- xdf_info[!is.na(xdf_info$Filter_rawsoni),]\n",
    "# v_id2Treename <- xdf_info$Treename\n",
    "# names(v_id2Treename) <- sapply(xdf_info$ID, function(x)substring(x,2))\n",
    "v_id2species <- xdf_info$new_Taxon_name\n",
    "names(v_id2species) <- sapply(xdf_info$ID, function(x)substring(x,2))\n",
    "species_in_order <- xdf_info[order(xdf_info$order_266mitoTree),]$ID\n",
    "species_in_order <- substr(species_in_order,2,length(species_in_order))\n",
    "\n",
    "\n",
    "oneline = df_files[1,]\n",
    "xdf_s <- read.csv(oneline$filename,header = FALSE,sep = '\\t', stringsAsFactors = FALSE)\n",
    "colnames(xdf_s) <- F.renameHead(colnames(xdf_s))\n",
    "xdf_s$species <- v_id2species[xdf_s$sample]\n",
    "#xdf_s$treename <- v_id2Treename[xdf_s$sample]\n",
    "xdf_s.clear <- gather(xdf_s,key = 'pop',value = 'fraction',-sample,-species)\n",
    "xdf_s.clear <- xdf_s.clear[order(xdf_s.clear$species,xdf_s.clear$pop,-xdf_s.clear$fraction),]\n",
    "\n",
    "p <- ggplot(xdf_s.clear,aes(x=sample,y=fraction,fill=pop)) + geom_bar(stat = 'identity') + scale_x_discrete(limits=species_in_order)\n",
    "\n",
    "\n",
    "\n",
    "F.plotOneline <- function(oneline,xtext=FALSE){\n",
    "  xdf_s <- read.csv(oneline$filename,header = FALSE,sep = '\\t', stringsAsFactors = FALSE)\n",
    "  colnames(xdf_s) <- F.renameHead(colnames(xdf_s))\n",
    "  xdf_s$species <- v_id2species[xdf_s$sample]\n",
    "  #xdf_s$treename <- v_id2Treename[xdf_s$sample]\n",
    "  xdf_s.clear <- gather(xdf_s,key = 'pop',value = 'fraction',-sample,-species)\n",
    "  xdf_s.clear <- xdf_s.clear[order(xdf_s.clear$species,xdf_s.clear$pop,-xdf_s.clear$fraction),]\n",
    "  \n",
    "  p <- ggplot(xdf_s.clear,aes(x=sample,y=fraction,fill=pop)) + geom_bar(stat = 'identity') + scale_x_discrete(limits=species_in_order[species_in_order %in% xdf_s$sample]) + ylab(paste0('seed',oneline$seed,'_',oneline$baseType,'_K=',oneline$population,\" \",oneline$splitID))\n",
    "  if (xtext) {\n",
    "    p <- p + geom_text(aes(x=sample,y=0,label=paste(sample,species)),angle=90,hjust=0)+ theme(legend.position=\"none\", axis.text.x= element_blank(), axis.title.x = element_blank())\n",
    "    #p <- p + theme(legend.position=\"none\", axis.text.x= element_text(angle = 90, hjust = 1), axis.title.x = element_blank())\n",
    "  }\n",
    "  else {\n",
    "    p <-  p + theme(legend.position=\"none\", axis.text.x= element_blank(), axis.title.x = element_blank())\n",
    "  }\n",
    "  return(p)\n",
    "}\n",
    "\n",
    "F.plotMultipleLines <- function(filelines,xtext=FALSE){\n",
    "  p <- list()\n",
    "  for (i in 1:nrow(filelines)){\n",
    "    p[[i]] <- F.plotOneline(filelines[i,],xtext = xtext)\n",
    "  }\n",
    "  #p[[i+1]] <- F.plotOneline(filelines[i+1,], xtext = TRUE)\n",
    "  return(grid.arrange(grobs=p,ncol=1))\n",
    "}\n",
    "\n",
    "#\"intron\",\"CDS\",\"mito\",\"wholeGap7\",\"zChr\"\n",
    "filelines = df_files[df_files$baseType == 'mito'& df_files$seed == 1000,]\n",
    "F.plotMultipleLines(filelines = filelines,xtext = TRUE)\n",
    "\n",
    "df_files2 = df_files\n",
    "df_files = df_files[df_files$population <= 6,]\n",
    "pdf('/home/xcao/w/2018Calephelis/20181231Structure/26/26rawsoni_seed3000.pdf',width = 8, height = 16)\n",
    "filelines = df_files[df_files$baseType == 'mito' & df_files$seed == 3000,]\n",
    "F.plotMultipleLines(filelines = filelines, xtext=TRUE)\n",
    "\n",
    "\n",
    "\n",
    "filelines = df_files[df_files$baseType == 'zChr'& df_files$seed == 3000,]\n",
    "F.plotMultipleLines(filelines = filelines, xtext=TRUE)\n",
    "\n",
    "filelines = df_files[df_files$baseType == 'CDS'& df_files$seed == 3000,]\n",
    "F.plotMultipleLines(filelines = filelines, xtext=TRUE)\n",
    "\n",
    "\n",
    "filelines = df_files[df_files$baseType == 'wholeGap7'& df_files$seed == 3000,]\n",
    "F.plotMultipleLines(filelines = filelines, xtext=TRUE)\n",
    "\n",
    "filelines = df_files[df_files$baseType == 'intron'& df_files$seed == 3000,]\n",
    "F.plotMultipleLines(filelines = filelines, xtext=TRUE)\n",
    "\n",
    "filelines = df_files[df_files$baseType == 'zChrCDS'& df_files$seed == 3000,]\n",
    "F.plotMultipleLines(filelines = filelines, xtext=TRUE)\n",
    "\n",
    "dev.off()\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## treemix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### treemix 35 C. rawsoni group"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### filter sites with non-gap ratio 0.8, basekeepMinCount 4 minBaseType 1\n",
    "\n",
    "```\n",
    "txt = 'python /home/xcao/p/xiaolongTools/utils/fastaAlignmentFilter.py -i {file_in} -o {file_out} -t 32 -b 4 -g 0.8 -B 1 &'\n",
    "workfolder = '/home/xcao/w/2018Calephelis/20190101rawsoni/Treemix/35groups'\n",
    "files = '''/home/xcao/w/2018Calephelis/20190101rawsoni/Calephelis35.CDS\n",
    "/home/xcao/w/2018Calephelis/20190101rawsoni/Calephelis35.intron\n",
    "/home/xcao/w/2018Calephelis/20190101rawsoni/Calephelis35.wholeGap7\n",
    "/home/xcao/w/2018Calephelis/20190101rawsoni/Calephelis35.zChr\n",
    "'''.split()\n",
    "import os\n",
    "for f in files:\n",
    "    basename = os.path.basename(f)\n",
    "    file_out = os.path.join(workfolder,basename+'.filter')\n",
    "    print(txt.format(file_in=f, file_out=file_out))\n",
    "    \n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### convert to Treemix input"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "txt = '''python3 /home/xcao/p/xiaolongTools/utils/fasta2TreemixInput_withGroups.py -i {file_in} -g {file_group} -o {file_out} -N 1 -G 0.8 &'''\n",
    "\n",
    "workfolder = '/home/xcao/w/2018Calephelis/20190101rawsoni/Treemix/35groups'\n",
    "files = '''/home/xcao/w/2018Calephelis/20190101rawsoni/Treemix/35groups/Calephelis35.CDS.filter\n",
    "/home/xcao/w/2018Calephelis/20190101rawsoni/Treemix/35groups/Calephelis35.intron.filter\n",
    "/home/xcao/w/2018Calephelis/20190101rawsoni/Treemix/35groups/Calephelis35.wholeGap7.filter\n",
    "/home/xcao/w/2018Calephelis/20190101rawsoni/Treemix/35groups/Calephelis35.zChr.filter\n",
    "'''.split()\n",
    "file_group = '/home/xcao/w/2018Calephelis/20190101rawsoni/Treemix/35groups/sample2group'\n",
    "\n",
    "import os\n",
    "for f in files:\n",
    "    basename = os.path.basename(f)\n",
    "    file_out = os.path.join(workfolder,basename+'.treemix')\n",
    "    os.system(txt.format(file_in=f, file_group=file_group, file_out=file_out))\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### run treemix\n",
    "```\n",
    "\n",
    "txt = '''/home/xcao/p/treemix/treemix -i {file_in} -o {file_out} -root outgroup &\n",
    "'''\n",
    "files = '''/home/xcao/w/2018Calephelis/20190101rawsoni/Treemix/35groups/Calephelis35.CDS.filter.treemix.gz\n",
    "/home/xcao/w/2018Calephelis/20190101rawsoni/Treemix/35groups/Calephelis35.intron.filter.treemix.gz\n",
    "/home/xcao/w/2018Calephelis/20190101rawsoni/Treemix/35groups/Calephelis35.wholeGap7.filter.treemix.gz\n",
    "/home/xcao/w/2018Calephelis/20190101rawsoni/Treemix/35groups/Calephelis35.zChr.filter.treemix.gz\n",
    "'''.split()\n",
    "\n",
    "import os\n",
    "for f in files:\n",
    "    file_out = '/home/xcao/w/2018Calephelis/20190101rawsoni/Treemix/35groups/result/' + os.path.basename(f)[:-3]\n",
    "    print(txt.format(file_in=f,file_out=file_out))\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### plot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "source('/home/xcao/p/treemix/treemix-1.13/src/plotting_funcs.R')\n",
    "library(glue)\n",
    "\n",
    "pdf('/home/xcao/w/2018Calephelis/20190101rawsoni/Treemix/35groups/Calephelis35.pdf',width = 8,height = 11)\n",
    "\n",
    "for (key in c('CDS','intron','wholeGap7','zChr')) {\n",
    "  #print(key)\n",
    "  filename.pre <- c('/home/xcao/w/2018Calephelis/20190101rawsoni/Treemix/35groups/result/Calephelis35.{key}.filter.treemix')\n",
    "\n",
    "    filename <- paste0(glue(filename.pre))\n",
    "    # print(filename)\n",
    "    plot_tree(filename,maintitle = paste0(key,' '))\n",
    "}\n",
    "\n",
    "dev.off()\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## careful study of Rawsoni"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Barcode and error detection"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### process and split barcodes from Nick\n",
    "```\n",
    "filename = '/home/xcao/w/2018Calephelis/summary/Calephelis_20181230_checked-barcodes.txt'\n",
    "\n",
    "ls = open(filename).readlines()\n",
    "ls = [e for e in ls if len(e)>1]\n",
    "\n",
    "workfolder = '/home/xcao/w/2018Calephelis/20190110BarcodeChecking/'\n",
    "import os\n",
    "os.chdir(workfolder)\n",
    "if not os.path.exists('barcodes'):\n",
    "    os.mkdir('barcodes')\n",
    "fout = open(workfolder+'id2notes','w')\n",
    "for e in ls:\n",
    "    annotation, seq = e.split()\n",
    "    if '|' not in annotation:\n",
    "        continue\n",
    "    sample_id = annotation.split('|')[1]\n",
    "    if sample_id.startswith('NVG-'):\n",
    "        sample_id = sample_id[4:]\n",
    "    fout.write(sample_id+'\\t'+annotation+'\\n')\n",
    "    open(workfolder+'barcodes/'+sample_id,'w').write('>'+sample_id+'\\n'+seq+'\\n')\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Extract and combine fastq to fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "import pandas as pd\n",
    "import os\n",
    "import glob\n",
    "\n",
    "folderBioHPC = '/work/archive/biophysics/Nick_lab/s185491/Celastrina/fq'\n",
    "folder_fq = '/archive/butterfly/ready_fastq/'\n",
    "folder_summary = '/home/xcao/w/2018Calephelis/summary/'\n",
    "\n",
    "file_summary = os.path.join(folder_summary,'20181217Calephelis.xlsx')\n",
    "df_summary = pd.read_excel(file_summary,sheet_name='Sheet1')\n",
    "df_summary = df_summary[~df_summary.PC1_35_CDS.isna()]\n",
    "sample_prefix = [e[1:] for e in df_summary['ID']]\n",
    "print('unique sample prefix', len(set(sample_prefix)))\n",
    "\n",
    "files_fq = glob.glob(folder_fq+'*')\n",
    "files_use = [e for e in files_fq if os.path.basename(e).split('_')[0] in sample_prefix]\n",
    "\n",
    "dc_fileuse = {}\n",
    "for e in files_use:\n",
    "    sample_id = os.path.basename(e).split('_')[0]\n",
    "    if sample_id not in dc_fileuse:\n",
    "        dc_fileuse[sample_id] = []\n",
    "    dc_fileuse[sample_id].append(e)\n",
    "\n",
    "from Bio import SeqIO\n",
    "outfolder = '/home/xcao/w/2018Calephelis/20190110BarcodeChecking/fa/'\n",
    "def fq2fa(k,files):\n",
    "    fout = open(outfolder+k,'w')\n",
    "    for f in files:\n",
    "        for seq in SeqIO.parse(f,'fastq'):\n",
    "            fout.write('>'+seq.id+'\\n'+str(seq.seq)+'\\n')\n",
    "    fout.close()\n",
    "from multiprocessing import Pool\n",
    "pool = Pool(32)\n",
    "pool.starmap(fq2fa, dc_fileuse.items())\n",
    "pool.close()\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### transfer to bioHPC\n",
    "```\n",
    "scp -r /home/xcao/w/2018Calephelis/20190110BarcodeChecking/ s185491@nucleus.biohpc.swmed.edu:/work/biophysics/s185491/20181226Calephelis/\n",
    " scp -r /home/xcao/w/2018Calephelis/20190110BarcodeChecking/fa/ s185491@nucleus.biohpc.swmed.edu:/work/biophysics/s185491/20181226Calephelis/20190110BarcodeChecking/\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### blast\n",
    "for each barcode, make a database with single seq. Blast with fasta input, first split to 32 parts, and combine the output"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Stop testing this method. too slow."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### jellyfish based barcode checking\n",
    "```\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# check existence of barcode files\n",
    "import os\n",
    "folder_barcode = '/work/biophysics/s185491/20181226Calephelis/20190110BarcodeChecking/barcodes/'\n",
    "folder_fa = '/work/biophysics/s185491/20181226Calephelis/20190110BarcodeChecking/fa/'\n",
    "fa_ids = os.listdir(folder_fa)\n",
    "barcode_ids = os.listdir(folder_barcode)\n",
    "sample_ids = open('/work/biophysics/s185491/20181226Calephelis/20190110BarcodeChecking/usingIDs').read().split()\n",
    "print(len([e for e in sample_ids if e in barcode_ids])) #barcode for all 35 using file exists\n",
    "\n",
    "\n",
    "#generate qsub scripts\n",
    "folder_result = '/work/biophysics/s185491/20181226Calephelis/20190110BarcodeChecking/result/'\n",
    "folder_fa = '/work/biophysics/s185491/20181226Calephelis/20190110BarcodeChecking/fa/'\n",
    "folder_cmds = '/work/biophysics/s185491/20181226Calephelis/20190110BarcodeChecking/cmds/'\n",
    "file_allbarcode = '/work/biophysics/s185491/20181226Calephelis/20190110BarcodeChecking/allBarcodes/intermediate_barcodes_20181231.fa'\n",
    "    \n",
    "qsub_txt = '''#!/bin/bash\n",
    "#SBATCH --job-name=barcode\n",
    "#SBATCH --partition=32GB\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=32\n",
    "#SBATCH --time=5-00:00:00\n",
    "#SBATCH --output=0.%j.out\n",
    "#SBATCH --mail-user=xiaolong.cao@utsouthwestern.edu\n",
    "#SBATCH --mail-type=ALL\n",
    "\n",
    "python /home2/s185491/p/xiaolongTools/utils/seq/checkBarcode.py -b {file_allbarcode} -i {folder_fa}/{sample_id} -T {sample_id} -o {folder_result}/{sample_id}\n",
    "\n",
    "'''\n",
    "\n",
    "for sample_id in sample_ids:\n",
    "    fout = open(folder_cmds+sample_id,'w')\n",
    "    fout.write(qsub_txt.format(folder_result=folder_result,  file_allbarcode=file_allbarcode, folder_fa=folder_fa,sample_id=sample_id))\n",
    "    fout.close()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### diagnostic characters of species"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### find genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "python /home/xcao/p/xiaolongTools/utils/speciesCharacter/characterFinderFromAlignment.py -a /home/xcao/w/2018Calephelis/20190101rawsoni/Calephelis35.CDS -l /home/xcao/w/genomes/Calephelis/3574_assembly_v1_withMito.alignment.exon.loc -A /home/xcao/w/genomes/Calephelis/cne_table_v1 -t /home/xcao/w/2018Calephelis/20190101rawsoni/20190128Feature/ExaML_result.Calephelis35.CDS.all5.sum.nwk.renamed -i /home/xcao/w/2018Calephelis/20190101rawsoni/20190128Feature/table_arizonensis -o /home/xcao/w/2018Calephelis/20190101rawsoni/20190128Feature/20190128Feature_arizonensis\n",
    "\n",
    "python /home/xcao/p/xiaolongTools/utils/speciesCharacter/characterFinderFromAlignment.py -a /home/xcao/w/2018Calephelis/20190101rawsoni/Calephelis35.CDS -l /home/xcao/w/genomes/Calephelis/3574_assembly_v1_withMito.alignment.exon.loc -A /home/xcao/w/genomes/Calephelis/cne_table_v1 -t /home/xcao/w/2018Calephelis/20190101rawsoni/20190128Feature/ExaML_result.Calephelis35.CDS.all5.sum.nwk.renamed -i /home/xcao/w/2018Calephelis/20190101rawsoni/20190128Feature/table_freemani -o /home/xcao/w/2018Calephelis/20190101rawsoni/20190128Feature/20190128Feature_freemani -S 0.8 -C 0.8 -D 0.7\n",
    "\n",
    "python /home/xcao/p/xiaolongTools/utils/speciesCharacter/characterFinderFromAlignment.py -a /home/xcao/w/2018Calephelis/20190101rawsoni/Calephelis35.CDS -l /home/xcao/w/genomes/Calephelis/3574_assembly_v1_withMito.alignment.exon.loc -A /home/xcao/w/genomes/Calephelis/cne_table_v1 -t /home/xcao/w/2018Calephelis/20190101rawsoni/20190128Feature/ExaML_result.Calephelis35.CDS.all5.sum.nwk.renamed -i /home/xcao/w/2018Calephelis/20190101rawsoni/20190128Feature/table_freemani_nuevoleon -o /home/xcao/w/2018Calephelis/20190101rawsoni/20190128Feature/20190128Feature_freemani_nuevoleon -S 0.8 -C 0.8 -D 0.8\n",
    "\n",
    "python /home/xcao/p/xiaolongTools/utils/speciesCharacter/characterFinderFromAlignment.py -a /home/xcao/w/2018Calephelis/20190101rawsoni/Calephelis35.CDS -l /home/xcao/w/genomes/Calephelis/3574_assembly_v1_withMito.alignment.exon.loc -A /home/xcao/w/genomes/Calephelis/cne_table_v1 -t /home/xcao/w/2018Calephelis/20190101rawsoni/20190128Feature/ExaML_result.Calephelis35.CDS.all5.sum.nwk.renamed -i /home/xcao/w/2018Calephelis/20190101rawsoni/20190128Feature/table_nuevoleon -o /home/xcao/w/2018Calephelis/20190101rawsoni/20190128Feature/20190128Feature_nuevoleon\n",
    "\n",
    "python /home/xcao/p/xiaolongTools/utils/speciesCharacter/characterFinderFromAlignment.py -a /home/xcao/w/2018Calephelis/20190101rawsoni/Calephelis35.CDS -l /home/xcao/w/genomes/Calephelis/3574_assembly_v1_withMito.alignment.exon.loc -A /home/xcao/w/genomes/Calephelis/cne_table_v1 -t /home/xcao/w/2018Calephelis/20190101rawsoni/20190128Feature/ExaML_result.Calephelis35.CDS.all5.sum.nwk.renamed -i /home/xcao/w/2018Calephelis/20190101rawsoni/20190128Feature/table_rawsoni -o /home/xcao/w/2018Calephelis/20190101rawsoni/20190128Feature/20190128Feature_rawsoni\n",
    "\n",
    "python /home/xcao/p/xiaolongTools/utils/speciesCharacter/characterFinderFromAlignment.py -a /home/xcao/w/2018Calephelis/20190101rawsoni/Calephelis35.CDS -l /home/xcao/w/genomes/Calephelis/3574_assembly_v1_withMito.alignment.exon.loc -A /home/xcao/w/genomes/Calephelis/cne_table_v1 -t /home/xcao/w/2018Calephelis/20190101rawsoni/20190128Feature/ExaML_result.Calephelis35.CDS.all5.sum.nwk.renamed -i /home/xcao/w/2018Calephelis/20190101rawsoni/20190128Feature/table_freemani_rawsoni -o /home/xcao/w/2018Calephelis/20190101rawsoni/20190128Feature/20190128Feature_freemani_rawsoni -S 0.8 -C 0.8 -D 0.8 -M 0.8\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### clean up 15109F05\n",
    "maybe contaminated by 15114F01_Arumecla_aruma"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Jellyfish 25mer count, inchworm Arumecla_aruma\n",
    "\n",
    "This method does not work. Inchworm cannot assemble the kmers well"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "keep contig longer than 100bp\n",
    "\n",
    "```\n",
    "jellyfish count -m 25 -s 2G -t 48 -C -o /home/xcao/w/2018Calephelis/20190207cleanup15109F05/Arumecla_aruma/Arumecla_aruma.25mer.jf /archive/butterfly/ready_fastq/15108F12* /archive/butterfly/ready_fastq/15114E11* /archive/butterfly/ready_fastq/15114E12* /archive/butterfly/ready_fastq/15114F01* /archive/butterfly/ready_fastq/15114F02* /archive/butterfly/ready_fastq/15114F03* /archive/butterfly/ready_fastq/15114F04*\n",
    "jellyfish dump /home/xcao/w/2018Calephelis/20190207cleanup15109F05/Arumecla_aruma/Arumecla_aruma.25mer.jf >/home/xcao/w/2018Calephelis/20190207cleanup15109F05/Arumecla_aruma/Arumecla_aruma.25mer.fa\n",
    "\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### kmer counting and sort by kmers\n",
    "use 21mer\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## study several groups\n",
    "20190208  \n",
    "* C.nemesis-group\n",
    "* C.nilus-group\n",
    "* C.velutina-group\n",
    "* C.burgeri\n",
    "* C.fulmen-group\n",
    "* C.rawsoni-group"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### check existance of files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### fq, map, stats\n",
    "```\n",
    "import pandas as pd\n",
    "import os\n",
    "import glob\n",
    "import itertools\n",
    "import sys\n",
    "sys.path.append('/home/xcao/p/xiaolongTools/utils/mapFile')\n",
    "import mapfileIO\n",
    "\n",
    "folder_fq = '/archive/butterfly/ready_fastq/'\n",
    "\n",
    "\n",
    "\n",
    "file_summary = '/work0/xcao/2018Calephelis/summary/20190207Calephelis.xlsx'\n",
    "df_summary = pd.read_excel(file_summary,sheet_name='Sheet1',dtype=str)\n",
    "sample_prefix = [e[1:] for e in df_summary['ID']]\n",
    "print('unique sample prefix', len(set(sample_prefix)))\n",
    "dc_sample2fq = {e:[] for e in sample_prefix}\n",
    "for k in dc_sample2fq:\n",
    "    fq = glob.glob('/archive/butterfly/ready_fastq/'+k+'_*')\n",
    "    dc_sample2fq[k] = fq\n",
    "dc_sample2map = {e:[] for e in sample_prefix}\n",
    "for k in dc_sample2map:\n",
    "    fmap = glob.glob('/archive/butterfly/maps/debiased/'+k+'_3574_assembly_v1_withMito_snp_step2.map')\n",
    "    dc_sample2map[k] = fmap\n",
    "dc_sample2stats = {e:[] for e in sample_prefix}\n",
    "for k in dc_sample2stats:\n",
    "    fstats = glob.glob('/archive/butterfly/unbias_pipeline_info/step4_postprocessing/final_stats/'+k+'_3574_assembly_v1_withMito_stat.report')\n",
    "    dc_sample2stats[k] = fstats\n",
    "sample_fq = [k for k,v in dc_sample2fq.items() if len(v)>0]\n",
    "sample_map = [k for k,v in dc_sample2map.items() if len(v)>0]\n",
    "sample_stats = [k for k,v in dc_sample2stats.items() if len(v)>0]\n",
    "print(len(sample_fq),'samples with fastq')\n",
    "print(len(sample_map), 'samples with map')\n",
    "print(len(sample_stats), 'samples with stats')\n",
    "samples_nomap = [e for e in sample_fq if e not in sample_map]\n",
    "print('\\n'.join(samples_nomap))\n",
    "\n",
    "#convert map files to npInt8\n",
    "dc_sample2np = {k:[] for k in dc_sample2map}\n",
    "for k in dc_sample2np:\n",
    "    file = glob.glob('/home/xcao/w/2018Calephelis/summary/map.npInt8.0/'+k+'_3574_assembly_v1_withMito_snp_step2.map')\n",
    "    dc_sample2np[k] = file\n",
    "\n",
    "file_map4convert = [v for k,v in dc_sample2map.items() if len(dc_sample2np[k]) == 0 and len(v)>0]\n",
    "file_map4convert = list(itertools.chain.from_iterable(file_map4convert))\n",
    "mapfileIO.read2Int8s(file_map4convert,strand=0, changeGap=True, threads=16, outfolder='/home/xcao/w/2018Calephelis/summary/map.npInt8.0/')\n",
    "mapfileIO.read2Int8s(file_map4convert,strand=2, changeGap=True, threads=16, outfolder='/home/xcao/w/2018Calephelis/summary/map.npInt8.2/')\n",
    "\n",
    "#for sample 3574, convert genome to mapfile\n",
    "file_genome = '/home/xcao/w/genomes/Calephelis/3574_assembly_v1_withMito.fa'\n",
    "from Bio import SeqIO\n",
    "l = []\n",
    "for s in SeqIO.parse(file_genome,'fasta'):\n",
    "    if s.id != '3574_mito':\n",
    "        l.append(str(s.seq))\n",
    "l = ''.join(l)\n",
    "fout = open('/home/xcao/w/2018Calephelis/summary/3574_3574_assembly_v1_withMito_snp_step2.map','w')\n",
    "fout.write('\\n'.join(list(l)))\n",
    "fout.write('\\n')\n",
    "fout.close()\n",
    "mapfileIO.read2Int8s(['/home/xcao/w/2018Calephelis/summary/3574_3574_assembly_v1_withMito_snp_step2.map'],strand=0, changeGap=True, threads=16, outfolder='/home/xcao/w/2018Calephelis/summary/map.npInt8.0/')\n",
    "sample_map.append('3574')\n",
    "\n",
    "\n",
    "file_summary = '/work/xcao/summary/All-sequenced.xlsx'\n",
    "sheets = '''C.nemesis-group\n",
    "C.nilus-group\n",
    "C.velutina-group\n",
    "C.burgeri\n",
    "C.fulmen-group\n",
    "C.rawsoni-group\n",
    "'''.split()\n",
    "\n",
    "dcdf = dict.fromkeys(sheets)\n",
    "for sheet_name in sheets:\n",
    "    df_summary = pd.read_excel(file_summary,sheet_name=sheet_name,dtype=str)\n",
    "    df_summary = df_summary[df_summary['DNA number'] != 'nan']\n",
    "    dcdf[sheet_name] = df_summary\n",
    "for sheet_name, df_summary in dcdf.items():\n",
    "    sample_prefix = [e.replace('NVG-','') for e in df_summary['DNA number']]\n",
    "    print(sheet_name, 'unique sample prefix', len(set(sample_prefix)))\n",
    "    sample_withmap = [e for e in sample_prefix if e in sample_map]\n",
    "    sample_withhoutmap = [e for e in sample_prefix if e not in sample_map]\n",
    "    print('samples with map file', len(sample_withmap))\n",
    "    print('samples without map file', sample_withhoutmap)\n",
    "    print('samples without fq', [e for e in sample_withhoutmap if e not in sample_fq])\n",
    "    fout = open('/home/xcao/w/2018Calephelis/summary/samples_'+sheet_name,'w')\n",
    "    fout.write('\\n'.join(sample_prefix))\n",
    "    fout.close()\n",
    "    \n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### result for 6 groups\n",
    "```\n",
    "C.nemesis-group unique sample prefix 32\n",
    "samples with map file 36\n",
    "samples without map file ['3751']\n",
    "samples without fq ['3751']\n",
    "C.nilus-group unique sample prefix 34\n",
    "samples with map file 34\n",
    "samples without map file []\n",
    "samples without fq []\n",
    "C.velutina-group unique sample prefix 16\n",
    "samples with map file 13\n",
    "samples without map file ['18043F05', '18098H01', '18098H02']\n",
    "samples without fq []\n",
    "C.burgeri unique sample prefix 9\n",
    "samples with map file 9\n",
    "samples without map file []\n",
    "samples without fq []\n",
    "C.fulmen-group unique sample prefix 35\n",
    "samples with map file 33\n",
    "samples without map file ['18098G10', '18098G12']\n",
    "samples without fq []\n",
    "C.rawsoni-group unique sample prefix 36\n",
    "samples with map file 35\n",
    "samples without map file ['18048A01']\n",
    "samples without fq []\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### get CDS sequences and zChrCDS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### all samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "import sys\n",
    "sys.path.append('/home2/s185491/p/xiaolongTools/utils/mapFile')\n",
    "import mapfileIO\n",
    "import glob\n",
    "\n",
    "folder = '/project/biophysics/Nick_lab/s185491/Calephelis/summary/map.npInt8.0/'\n",
    "files = glob.glob(folder+'*')\n",
    "print('total map npInt8 files', len(files))\n",
    "\n",
    "file_filter = '/project/biophysics/Nick_lab/s185491/Calephelis/summary/map.sitesInfo/gff.0inter.2intron.6CDS'\n",
    "outfile = '/project/biophysics/Nick_lab/s185491/Calephelis/phylo/allseq/Calephelis270.CDS'\n",
    "mapfileIO.npInt8stoFastas(files, threads=48, output=outfile, headerFun=lambda x:x.split('_')[0], npFilter=file_filter, npFilterFun=lambda x:x==6, merged=True)\n",
    "\n",
    "file_filter = '/project/biophysics/Nick_lab/s185491/Calephelis/summary/map.sitesInfo/zChrCDSFilter'\n",
    "outfile = '/project/biophysics/Nick_lab/s185491/Calephelis/phylo/allseq/Calephelis270.zChrCDS'\n",
    "mapfileIO.npInt8stoFastas(files, threads=32, output=outfile, headerFun=lambda x:x.split('_')[0], npFilter=file_filter, npFilterFun=None, merged=True)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### get sequences for each group and split CDS 100 parts, zChrCDS 5 parts\n",
    "```\n",
    "file_CDS = '/project/biophysics/Nick_lab/s185491/Calephelis/phylo/allseq/Calephelis270.CDS'\n",
    "file_zChrCDS = '/project/biophysics/Nick_lab/s185491/Calephelis/phylo/allseq/Calephelis270.zChrCDS'\n",
    "groups = '''C.nemesis-group\n",
    "C.nilus-group\n",
    "C.velutina-group\n",
    "C.burgeri\n",
    "C.fulmen-group\n",
    "C.rawsoni-group\n",
    "'''.split()\n",
    "\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "seqs_CDS = list(SeqIO.parse(file_CDS,'fasta'))\n",
    "seqs_zChrCDS = list(SeqIO.parse(file_zChrCDS,'fasta'))\n",
    "for g in groups:\n",
    "    outfolder = '/project/biophysics/Nick_lab/s185491/Calephelis/phylo/'+g\n",
    "    if not os.path.exists(outfolder):\n",
    "        os.makedirs(outfolder)\n",
    "    samples = open('/project/biophysics/Nick_lab/s185491/Calephelis/summary/samples_'+g).read().split()\n",
    "    with open(os.path.join(outfolder,'CDS.fa'),'w') as f:\n",
    "        for s in seqs_CDS:\n",
    "            if s.id in samples:\n",
    "                f.write('>' + s.id +'\\n'+str(s.seq)+'\\n')\n",
    "    with open(os.path.join(outfolder,'zChrCDS.fa'),'w') as f:\n",
    "        for s in seqs_zChrCDS:\n",
    "            if s.id in samples:\n",
    "                f.write('>' + s.id +'\\n'+str(s.seq)+'\\n')\n",
    "for g in groups:\n",
    "    outfolder = '/project/biophysics/Nick_lab/s185491/Calephelis/phylo/'+g\n",
    "    cmd = 'cd {outfolder} && python /home2/s185491/p/xiaolongTools/utils/splitFastaAlignments2NPartsContinuingEqualStepOrRandom.py -i CDS.fa -m 1 -t 48 -b 10 -N 100'.format(outfolder=outfolder)\n",
    "    os.system(cmd)\n",
    "    cmd = 'cd {outfolder} && python /home2/s185491/p/xiaolongTools/utils/splitFastaAlignments2NPartsContinuingEqualStepOrRandom.py -i zChrCDS.fa -m 1 -t 48 -b 10 -N 5'.format(outfolder=outfolder)\n",
    "    os.system(cmd)\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### run ExaML"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "txt_qsub = '''#!/bin/bash\n",
    "#SBATCH --job-name={n}\n",
    "#SBATCH --partition=32GB\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=32\n",
    "#SBATCH --time=5-00:00:00\n",
    "#SBATCH --output=0.%j.out\n",
    "####SBATCH --mail-user=xiaolong.cao@utsouthwestern.edu\n",
    "#SBATCH --mail-type=ALL\n",
    "\n",
    "module load openmpi/intel/3.1.1\n",
    "\n",
    "python3 /home2/s185491/p/xiaolongTools/utils/runExaML_withFastaForLargeFiles.py -i {filename} -T 32\n",
    "'''\n",
    "\n",
    "import glob\n",
    "files = glob.glob('/project/biophysics/Nick_lab/s185491/Calephelis/phylo/*/*.fa.split*')\n",
    "\n",
    "import os\n",
    "for filename in files:\n",
    "    basename = os.path.basename(filename)\n",
    "    groupname = os.path.basename(os.path.dirname(filename))\n",
    "    fout = open('/project/biophysics/Nick_lab/s185491/Calephelis/phylo/qsub/ExaML'+groupname+basename,'w')\n",
    "    fout.write(txt_qsub.format(filename=filename,n=basename))\n",
    "    fout.close()\n",
    "\n",
    "groups = '''C.nemesis-group\n",
    "C.nilus-group\n",
    "C.velutina-group\n",
    "C.burgeri\n",
    "C.fulmen-group\n",
    "C.rawsoni-group\n",
    "'''.split()\n",
    "\n",
    "import os\n",
    "for g in groups:\n",
    "    outfolder = '/project/biophysics/Nick_lab/s185491/Calephelis/phylo/'+g\n",
    "    cmd = '''cd {outfolder} && \\\n",
    "    cat ExaML_result.CDS.fa.split100.* >ExaML_result.CDS.all &&\\\n",
    "    /home2/s185491/p/anaconda3/anaconda520/bin/sumtrees.py -f 0.00 ExaML_result.CDS.all>ExaML_result.CDS.sum &&\\\n",
    "    python /home2/s185491/p/xiaolongTools/utils/trees/sumtreesNEXUS2newick.py -i ExaML_result.CDS.sum &&\\\n",
    "    /home2/mtang/anaconda2/bin/python /work/biophysics/mtang/SNP_calling/scripts/rename_tree_fullname.py  ExaML_result.CDS.sum.nwk &&\\\n",
    "    java -Xmx3000m -jar /home2/s185491/p/Astral/astral.5.6.2.jar -i ExaML_result.CDS.all -o ExaML_result.CDS.astral &&\\\n",
    "    /home2/mtang/anaconda2/bin/python /work/biophysics/mtang/SNP_calling/scripts/rename_tree_fullname.py  ExaML_result.CDS.astral '''.format(outfolder=outfolder)\n",
    "    os.system(cmd)\n",
    "\n",
    "for g in groups:\n",
    "    outfolder = '/project/biophysics/Nick_lab/s185491/Calephelis/phylo/'+g\n",
    "    cmd = '''cd {outfolder} && \\\n",
    "    cat ExaML_result.zChrCDS.fa.split5.* >ExaML_result.zChrCDS.all &&\\\n",
    "    /home2/s185491/p/anaconda3/anaconda520/bin/sumtrees.py -f 0.00 ExaML_result.zChrCDS.all>ExaML_result.zChrCDS.sum &&\\\n",
    "    python /home2/s185491/p/xiaolongTools/utils/trees/sumtreesNEXUS2newick.py -i ExaML_result.zChrCDS.sum &&\\\n",
    "    /home2/mtang/anaconda2/bin/python /work/biophysics/mtang/SNP_calling/scripts/rename_tree_fullname.py  ExaML_result.zChrCDS.sum.nwk &&\\\n",
    "    java -Xmx3000m -jar /home2/s185491/p/Astral/astral.5.6.2.jar -i ExaML_result.zChrCDS.all -o ExaML_result.zChrCDS.astral &&\\\n",
    "    /home2/mtang/anaconda2/bin/python /work/biophysics/mtang/SNP_calling/scripts/rename_tree_fullname.py  ExaML_result.zChrCDS.astral '''.format(outfolder=outfolder)\n",
    "    os.system(cmd)\n",
    "\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": true,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "384px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
